{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#DATA_FOLDER = \"/home/vinz/Desktop/ADA/ADA2017-Tutorials/02 - Intro to Pandas/Data\" # Use the data folder provided in Tutorial 02 - Intro to Pandas.\n",
    "DATA_FOLDER = \"./Data\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average* per year of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strategy\n",
    "\n",
    "We will import and wrangle the data for each country separately and then combine it into a single data frame.\n",
    "\n",
    "As the data files already contain variables with nation wide statistics, we will use these values instead of manually aggregating the data city by city."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def ebola_csv_iter(folder):\n",
    "    \"\"\"Utility function returning an iterator over the csv files in given folder\"\"\"\n",
    "    path = ''.join([DATA_FOLDER, '/ebola/', folder, '/'])\n",
    "    for file in os.listdir(path):\n",
    "        if file.endswith(\".csv\"):\n",
    "            yield file, ''.join([path, file])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Guinea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dfs = {}\n",
    "tables_with_duplicates = []\n",
    "for file, path in ebola_csv_iter('guinea_data'):\n",
    "    df = pd.read_csv(path,\n",
    "                     parse_dates=['Date'],\n",
    "                     usecols=['Date', 'Description', 'Totals'])\n",
    "    df.rename(columns={'Description': 'Variable', 'Totals': 'National'}, inplace=True)\n",
    "    df.Variable = df.Variable.astype('category')\n",
    "    if len(df.Variable.unique()) != len(df.Variable): # Check for duplicates\n",
    "        tables_with_duplicates.append(file)\n",
    "    dfs[file] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tables_with_duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As there are no duplicates, we can pivot all the dataframes to have a timeseries dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = [df.pivot(index='Date', columns='Variable', values='National') for df in dfs.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Variable</th>\n",
       "      <th>Cumulative (confirmed + probable + suspects)</th>\n",
       "      <th>Fatality rate for confirmed and probables</th>\n",
       "      <th>New admits to CTE so far</th>\n",
       "      <th>New cases of confirmed</th>\n",
       "      <th>New cases of confirmed among health workers</th>\n",
       "      <th>New cases of probables</th>\n",
       "      <th>New cases of suspects</th>\n",
       "      <th>New contacts registered so far</th>\n",
       "      <th>New deaths registered</th>\n",
       "      <th>New deaths registered among health workers</th>\n",
       "      <th>...</th>\n",
       "      <th>Total deaths of confirmed</th>\n",
       "      <th>Total deaths of probables</th>\n",
       "      <th>Total deaths of suspects</th>\n",
       "      <th>Total deaths registered among health workers</th>\n",
       "      <th>Total new cases registered so far</th>\n",
       "      <th>Total number of admissions to CTE</th>\n",
       "      <th>Total number of exits from CTE</th>\n",
       "      <th>Total number of hospitalized cases in CTE</th>\n",
       "      <th>Total samples tested</th>\n",
       "      <th>Total suspected non-class cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-09-02</th>\n",
       "      <td>790</td>\n",
       "      <td>63%</td>\n",
       "      <td>29</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>349</td>\n",
       "      <td>150</td>\n",
       "      <td>2</td>\n",
       "      <td>26</td>\n",
       "      <td>25</td>\n",
       "      <td>1144</td>\n",
       "      <td>24</td>\n",
       "      <td>105</td>\n",
       "      <td>2018</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Variable   Cumulative (confirmed + probable + suspects)  \\\n",
       "Date                                                      \n",
       "2014-09-02                                          790   \n",
       "\n",
       "Variable   Fatality rate for confirmed and probables New admits to CTE so far  \\\n",
       "Date                                                                            \n",
       "2014-09-02                                       63%                       29   \n",
       "\n",
       "Variable   New cases of confirmed New cases of confirmed among health workers  \\\n",
       "Date                                                                            \n",
       "2014-09-02                     14                                           0   \n",
       "\n",
       "Variable   New cases of probables New cases of suspects  \\\n",
       "Date                                                      \n",
       "2014-09-02                      0                    11   \n",
       "\n",
       "Variable   New contacts registered so far New deaths registered  \\\n",
       "Date                                                              \n",
       "2014-09-02                              0                     5   \n",
       "\n",
       "Variable   New deaths registered among health workers  \\\n",
       "Date                                                    \n",
       "2014-09-02                                          0   \n",
       "\n",
       "Variable                 ...               Total deaths of confirmed  \\\n",
       "Date                     ...                                           \n",
       "2014-09-02               ...                                     349   \n",
       "\n",
       "Variable   Total deaths of probables Total deaths of suspects  \\\n",
       "Date                                                            \n",
       "2014-09-02                       150                        2   \n",
       "\n",
       "Variable   Total deaths registered among health workers  \\\n",
       "Date                                                      \n",
       "2014-09-02                                           26   \n",
       "\n",
       "Variable   Total new cases registered so far  \\\n",
       "Date                                           \n",
       "2014-09-02                                25   \n",
       "\n",
       "Variable   Total number of admissions to CTE Total number of exits from CTE  \\\n",
       "Date                                                                          \n",
       "2014-09-02                              1144                             24   \n",
       "\n",
       "Variable   Total number of hospitalized cases in CTE Total samples tested  \\\n",
       "Date                                                                        \n",
       "2014-09-02                                       105                 2018   \n",
       "\n",
       "Variable   Total suspected non-class cases  \n",
       "Date                                        \n",
       "2014-09-02                             642  \n",
       "\n",
       "[1 rows x 32 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "guinea = pd.concat(dfs).sort_index() # Create the data series: index=date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Variable</th>\n",
       "      <th>Cumulative (confirmed + probable + suspects)</th>\n",
       "      <th>Fatality rate for confirmed and probables</th>\n",
       "      <th>New admits to CTE so far</th>\n",
       "      <th>New cases of confirmed</th>\n",
       "      <th>New cases of confirmed among health workers</th>\n",
       "      <th>New cases of probables</th>\n",
       "      <th>New cases of suspects</th>\n",
       "      <th>New contacts registered so far</th>\n",
       "      <th>New deaths registered</th>\n",
       "      <th>New deaths registered among health workers</th>\n",
       "      <th>...</th>\n",
       "      <th>Total new cases registered so far</th>\n",
       "      <th>Total number of admissions to CTE</th>\n",
       "      <th>Total number of exits from CTE</th>\n",
       "      <th>Total number of female cases</th>\n",
       "      <th>Total number of hospitalized cases in CTE</th>\n",
       "      <th>Total number of male cases</th>\n",
       "      <th>Total of cured in confirmed cases in CTE</th>\n",
       "      <th>Total of deaths in confirmed cases in CTE</th>\n",
       "      <th>Total samples tested</th>\n",
       "      <th>Total suspected non-class cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-08-04</th>\n",
       "      <td>495</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>277.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>220.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>138.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-26</th>\n",
       "      <td>661</td>\n",
       "      <td>66%</td>\n",
       "      <td>22</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18</td>\n",
       "      <td>137</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>997</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1700</td>\n",
       "      <td>518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-27</th>\n",
       "      <td>663</td>\n",
       "      <td>66%</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>127</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>999</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1719</td>\n",
       "      <td>518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-30</th>\n",
       "      <td>707</td>\n",
       "      <td>66%</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>21</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>1094</td>\n",
       "      <td>33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>82</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1858</td>\n",
       "      <td>560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-31</th>\n",
       "      <td>749</td>\n",
       "      <td>65%</td>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>1113</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1941</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-02</th>\n",
       "      <td>790</td>\n",
       "      <td>63%</td>\n",
       "      <td>29</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>1144</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-04</th>\n",
       "      <td>823</td>\n",
       "      <td>61%</td>\n",
       "      <td>13</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>21</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>30</td>\n",
       "      <td>1191</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>103</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>249</td>\n",
       "      <td>682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-07</th>\n",
       "      <td>861</td>\n",
       "      <td>65%</td>\n",
       "      <td>50</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>1256</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2236</td>\n",
       "      <td>714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-08</th>\n",
       "      <td>861</td>\n",
       "      <td>65%</td>\n",
       "      <td>50</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>1256</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2236</td>\n",
       "      <td>714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-09</th>\n",
       "      <td>867</td>\n",
       "      <td>65%</td>\n",
       "      <td>52</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>1272</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>92</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2274</td>\n",
       "      <td>747</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Variable   Cumulative (confirmed + probable + suspects)  \\\n",
       "Date                                                      \n",
       "2014-08-04                                          495   \n",
       "2014-08-26                                          661   \n",
       "2014-08-27                                          663   \n",
       "2014-08-30                                          707   \n",
       "2014-08-31                                          749   \n",
       "2014-09-02                                          790   \n",
       "2014-09-04                                          823   \n",
       "2014-09-07                                          861   \n",
       "2014-09-08                                          861   \n",
       "2014-09-09                                          867   \n",
       "\n",
       "Variable   Fatality rate for confirmed and probables New admits to CTE so far  \\\n",
       "Date                                                                            \n",
       "2014-08-04                                       NaN                      NaN   \n",
       "2014-08-26                                       66%                       22   \n",
       "2014-08-27                                       66%                       12   \n",
       "2014-08-30                                       66%                        9   \n",
       "2014-08-31                                       65%                       26   \n",
       "2014-09-02                                       63%                       29   \n",
       "2014-09-04                                       61%                       13   \n",
       "2014-09-07                                       65%                       50   \n",
       "2014-09-08                                       65%                       50   \n",
       "2014-09-09                                       65%                       52   \n",
       "\n",
       "Variable   New cases of confirmed New cases of confirmed among health workers  \\\n",
       "Date                                                                            \n",
       "2014-08-04                      4                                         NaN   \n",
       "2014-08-26                     10                                           0   \n",
       "2014-08-27                     10                                           0   \n",
       "2014-08-30                      9                                           0   \n",
       "2014-08-31                     29                                           0   \n",
       "2014-09-02                     14                                           0   \n",
       "2014-09-04                     17                                           0   \n",
       "2014-09-07                     11                                           2   \n",
       "2014-09-08                     11                                           2   \n",
       "2014-09-09                      7                                           0   \n",
       "\n",
       "Variable   New cases of probables New cases of suspects  \\\n",
       "Date                                                      \n",
       "2014-08-04                      0                     5   \n",
       "2014-08-26                    NaN                    18   \n",
       "2014-08-27                      0                    12   \n",
       "2014-08-30                      0                    15   \n",
       "2014-08-31                      8                     9   \n",
       "2014-09-02                      0                    11   \n",
       "2014-09-04                      0                    13   \n",
       "2014-09-07                      0                     5   \n",
       "2014-09-08                      0                     5   \n",
       "2014-09-09                      0                     9   \n",
       "\n",
       "Variable   New contacts registered so far New deaths registered  \\\n",
       "Date                                                              \n",
       "2014-08-04                            NaN                   NaN   \n",
       "2014-08-26                            137                     5   \n",
       "2014-08-27                            127                     2   \n",
       "2014-08-30                             21                     5   \n",
       "2014-08-31                              0                     3   \n",
       "2014-09-02                              0                     5   \n",
       "2014-09-04                             21                     5   \n",
       "2014-09-07                              5                     4   \n",
       "2014-09-08                              5                     4   \n",
       "2014-09-09                              1                     7   \n",
       "\n",
       "Variable   New deaths registered among health workers  \\\n",
       "Date                                                    \n",
       "2014-08-04                                        NaN   \n",
       "2014-08-26                                          0   \n",
       "2014-08-27                                          0   \n",
       "2014-08-30                                          0   \n",
       "2014-08-31                                          0   \n",
       "2014-09-02                                          0   \n",
       "2014-09-04                                          0   \n",
       "2014-09-07                                          0   \n",
       "2014-09-08                                          0   \n",
       "2014-09-09                                          0   \n",
       "\n",
       "Variable                 ...                 \\\n",
       "Date                     ...                  \n",
       "2014-08-04               ...                  \n",
       "2014-08-26               ...                  \n",
       "2014-08-27               ...                  \n",
       "2014-08-30               ...                  \n",
       "2014-08-31               ...                  \n",
       "2014-09-02               ...                  \n",
       "2014-09-04               ...                  \n",
       "2014-09-07               ...                  \n",
       "2014-09-08               ...                  \n",
       "2014-09-09               ...                  \n",
       "\n",
       "Variable    Total new cases registered so far  \\\n",
       "Date                                            \n",
       "2014-08-04                                  9   \n",
       "2014-08-26                                 28   \n",
       "2014-08-27                                 22   \n",
       "2014-08-30                                 24   \n",
       "2014-08-31                                 46   \n",
       "2014-09-02                                 25   \n",
       "2014-09-04                                 30   \n",
       "2014-09-07                                 16   \n",
       "2014-09-08                                 16   \n",
       "2014-09-09                                 16   \n",
       "\n",
       "Variable    Total number of admissions to CTE  Total number of exits from CTE  \\\n",
       "Date                                                                            \n",
       "2014-08-04                                NaN                             NaN   \n",
       "2014-08-26                                997                               6   \n",
       "2014-08-27                                999                              14   \n",
       "2014-08-30                               1094                              33   \n",
       "2014-08-31                               1113                              10   \n",
       "2014-09-02                               1144                              24   \n",
       "2014-09-04                               1191                              15   \n",
       "2014-09-07                               1256                               8   \n",
       "2014-09-08                               1256                               8   \n",
       "2014-09-09                               1272                              21   \n",
       "\n",
       "Variable    Total number of female cases  \\\n",
       "Date                                       \n",
       "2014-08-04                         277.0   \n",
       "2014-08-26                           NaN   \n",
       "2014-08-27                           NaN   \n",
       "2014-08-30                           NaN   \n",
       "2014-08-31                           NaN   \n",
       "2014-09-02                           NaN   \n",
       "2014-09-04                           NaN   \n",
       "2014-09-07                           NaN   \n",
       "2014-09-08                           NaN   \n",
       "2014-09-09                           NaN   \n",
       "\n",
       "Variable    Total number of hospitalized cases in CTE  \\\n",
       "Date                                                    \n",
       "2014-08-04                                        NaN   \n",
       "2014-08-26                                         87   \n",
       "2014-08-27                                         80   \n",
       "2014-08-30                                         82   \n",
       "2014-08-31                                         99   \n",
       "2014-09-02                                        105   \n",
       "2014-09-04                                        103   \n",
       "2014-09-07                                        102   \n",
       "2014-09-08                                        102   \n",
       "2014-09-09                                         92   \n",
       "\n",
       "Variable   Total number of male cases  \\\n",
       "Date                                    \n",
       "2014-08-04                      220.0   \n",
       "2014-08-26                        NaN   \n",
       "2014-08-27                        NaN   \n",
       "2014-08-30                        NaN   \n",
       "2014-08-31                        NaN   \n",
       "2014-09-02                        NaN   \n",
       "2014-09-04                        NaN   \n",
       "2014-09-07                        NaN   \n",
       "2014-09-08                        NaN   \n",
       "2014-09-09                        NaN   \n",
       "\n",
       "Variable    Total of cured in confirmed cases in CTE  \\\n",
       "Date                                                   \n",
       "2014-08-04                                      89.0   \n",
       "2014-08-26                                       NaN   \n",
       "2014-08-27                                       NaN   \n",
       "2014-08-30                                       NaN   \n",
       "2014-08-31                                       NaN   \n",
       "2014-09-02                                       NaN   \n",
       "2014-09-04                                       NaN   \n",
       "2014-09-07                                       NaN   \n",
       "2014-09-08                                       NaN   \n",
       "2014-09-09                                       NaN   \n",
       "\n",
       "Variable   Total of deaths in confirmed cases in CTE Total samples tested  \\\n",
       "Date                                                                        \n",
       "2014-08-04                                     138.0                  NaN   \n",
       "2014-08-26                                       NaN                 1700   \n",
       "2014-08-27                                       NaN                 1719   \n",
       "2014-08-30                                       NaN                 1858   \n",
       "2014-08-31                                       NaN                 1941   \n",
       "2014-09-02                                       NaN                 2018   \n",
       "2014-09-04                                       NaN                  249   \n",
       "2014-09-07                                       NaN                 2236   \n",
       "2014-09-08                                       NaN                 2236   \n",
       "2014-09-09                                       NaN                 2274   \n",
       "\n",
       "Variable    Total suspected non-class cases  \n",
       "Date                                         \n",
       "2014-08-04                              NaN  \n",
       "2014-08-26                              518  \n",
       "2014-08-27                              518  \n",
       "2014-08-30                              560  \n",
       "2014-08-31                              642  \n",
       "2014-09-02                              642  \n",
       "2014-09-04                              682  \n",
       "2014-09-07                              714  \n",
       "2014-09-08                              714  \n",
       "2014-09-09                              747  \n",
       "\n",
       "[10 rows x 60 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guinea.head(n=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Liberia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = {}\n",
    "tables_with_duplicates = []\n",
    "\n",
    "for file, path in ebola_csv_iter('liberia_data'):\n",
    "    df = pd.read_csv(path,\n",
    "                     parse_dates=['Date'],\n",
    "                     usecols=['Date', 'Variable', 'National'])\n",
    "    df.Variable = df.Variable.astype('category')\n",
    "        \n",
    "    if len(df.Variable.unique()) != len(df.Variable): # Check for duplicates\n",
    "        tables_with_duplicates.append(file)\n",
    "            \n",
    "    dfs[file] = df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We remark that only a data file contains duplicate variables for the same day. Before dropping the duplicates, we look at the duplicated lines:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['2014-10-04-v142.csv']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tables_with_duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Variable</th>\n",
       "      <th>National</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Cumulative (confirmed + probable + suspected)</td>\n",
       "      <td>3921.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Cumulative cases among HCW</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Cumulative deaths among HCW</td>\n",
       "      <td>92.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Total death/s in confirmed cases</td>\n",
       "      <td>1012.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Total death/s in probable cases</td>\n",
       "      <td>699.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Total death/s in suspected cases</td>\n",
       "      <td>488.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Total death/s in confirmed, probable, suspecte...</td>\n",
       "      <td>2199.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Cumulative (confirmed + probable + suspected)</td>\n",
       "      <td>3929.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Cumulative cases among HCW</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Cumulative deaths among HCW</td>\n",
       "      <td>94.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Total death/s in confirmed cases</td>\n",
       "      <td>1018.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Total death/s in probable cases</td>\n",
       "      <td>701.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Total death/s in suspected cases</td>\n",
       "      <td>491.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>2014-10-04</td>\n",
       "      <td>Total death/s in confirmed, probable, suspecte...</td>\n",
       "      <td>2210.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Date                                           Variable  National\n",
       "8  2014-10-04      Cumulative (confirmed + probable + suspected)    3921.0\n",
       "10 2014-10-04                         Cumulative cases among HCW     190.0\n",
       "12 2014-10-04                        Cumulative deaths among HCW      92.0\n",
       "17 2014-10-04                   Total death/s in confirmed cases    1012.0\n",
       "18 2014-10-04                    Total death/s in probable cases     699.0\n",
       "19 2014-10-04                   Total death/s in suspected cases     488.0\n",
       "20 2014-10-04  Total death/s in confirmed, probable, suspecte...    2199.0\n",
       "31 2014-10-04      Cumulative (confirmed + probable + suspected)    3929.0\n",
       "32 2014-10-04                         Cumulative cases among HCW     192.0\n",
       "33 2014-10-04                        Cumulative deaths among HCW      94.0\n",
       "34 2014-10-04                   Total death/s in confirmed cases    1018.0\n",
       "35 2014-10-04                    Total death/s in probable cases     701.0\n",
       "36 2014-10-04                   Total death/s in suspected cases     491.0\n",
       "37 2014-10-04  Total death/s in confirmed, probable, suspecte...    2210.0"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with_duplicates = dfs[tables_with_duplicates[0]]\n",
    "with_duplicates[with_duplicates.duplicated(keep=False, subset='Variable')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "7 Variables are repeated twice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "first = with_duplicates[with_duplicates.duplicated(keep='last', subset='Variable')][['Variable', 'National']]\n",
    "second = with_duplicates[with_duplicates.duplicated(keep='first', subset='Variable')][['Variable', 'National']]\n",
    "first = first.set_index('Variable')\n",
    "second = second.set_index('Variable')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>National</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variable</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cumulative (confirmed + probable + suspected)</th>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cumulative cases among HCW</th>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cumulative deaths among HCW</th>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total death/s in confirmed cases</th>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total death/s in confirmed, probable, suspected cases</th>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    National\n",
       "Variable                                                    \n",
       "Cumulative (confirmed + probable + suspected)            8.0\n",
       "Cumulative cases among HCW                               2.0\n",
       "Cumulative deaths among HCW                              2.0\n",
       "Total death/s in confirmed cases                         6.0\n",
       "Total death/s in probable cases                          2.0\n",
       "Total death/s in suspected cases                         3.0\n",
       "Total death/s in confirmed, probable, suspected...      11.0"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "second - first"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We decided to keep the bigger values as the difference is not very relevant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with_duplicates.drop_duplicates(subset='Variable', keep='last', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = [df.pivot(index='Date', columns='Variable', values='National') for df in dfs.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Variable</th>\n",
       "      <th>Case Fatality Rate (CFR) - Confirmed &amp; Probable Cases</th>\n",
       "      <th>Contacts lost to follow-up</th>\n",
       "      <th>Contacts seen</th>\n",
       "      <th>Contacts who completed 21 day follow-up</th>\n",
       "      <th>Cumulative admission/isolation</th>\n",
       "      <th>Cumulative cases among HCW</th>\n",
       "      <th>Cumulative deaths among HCW</th>\n",
       "      <th>Currently under follow-up</th>\n",
       "      <th>New Case/s (Probable)</th>\n",
       "      <th>New Case/s (Suspected)</th>\n",
       "      <th>...</th>\n",
       "      <th>Total contacts listed</th>\n",
       "      <th>Total death/s in confirmed cases</th>\n",
       "      <th>Total death/s in confirmed, probable, suspected cases</th>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "      <th>Total discharges</th>\n",
       "      <th>Total no. currently in Treatment Units</th>\n",
       "      <th>Total probable cases</th>\n",
       "      <th>Total specimens tested</th>\n",
       "      <th>Total suspected cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-09-10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3527.0</td>\n",
       "      <td>321.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>3583.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7081.0</td>\n",
       "      <td>563.0</td>\n",
       "      <td>1296.0</td>\n",
       "      <td>472.0</td>\n",
       "      <td>261.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>1078.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>539.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Variable    Case Fatality Rate (CFR) - Confirmed & Probable Cases  \\\n",
       "Date                                                                \n",
       "2014-09-10                                                NaN       \n",
       "\n",
       "Variable    Contacts lost to follow-up  Contacts seen  \\\n",
       "Date                                                    \n",
       "2014-09-10                         3.0         3527.0   \n",
       "\n",
       "Variable    Contacts who completed 21 day follow-up  \\\n",
       "Date                                                  \n",
       "2014-09-10                                    321.0   \n",
       "\n",
       "Variable    Cumulative admission/isolation  Cumulative cases among HCW  \\\n",
       "Date                                                                     \n",
       "2014-09-10                            49.0                       169.0   \n",
       "\n",
       "Variable    Cumulative deaths among HCW  Currently under follow-up  \\\n",
       "Date                                                                 \n",
       "2014-09-10                         82.0                     3583.0   \n",
       "\n",
       "Variable    New Case/s (Probable)  New Case/s (Suspected)  \\\n",
       "Date                                                        \n",
       "2014-09-10                   28.0                    29.0   \n",
       "\n",
       "Variable            ...            Total contacts listed  \\\n",
       "Date                ...                                    \n",
       "2014-09-10          ...                           7081.0   \n",
       "\n",
       "Variable    Total death/s in confirmed cases  \\\n",
       "Date                                           \n",
       "2014-09-10                             563.0   \n",
       "\n",
       "Variable    Total death/s in confirmed, probable, suspected cases  \\\n",
       "Date                                                                \n",
       "2014-09-10                                             1296.0       \n",
       "\n",
       "Variable    Total death/s in probable cases  Total death/s in suspected cases  \\\n",
       "Date                                                                            \n",
       "2014-09-10                            472.0                             261.0   \n",
       "\n",
       "Variable    Total discharges  Total no. currently in Treatment Units  \\\n",
       "Date                                                                   \n",
       "2014-09-10               2.0                                    49.0   \n",
       "\n",
       "Variable    Total probable cases  Total specimens tested  \\\n",
       "Date                                                       \n",
       "2014-09-10                1078.0                    32.0   \n",
       "\n",
       "Variable    Total suspected cases  \n",
       "Date                               \n",
       "2014-09-10                  539.0  \n",
       "\n",
       "[1 rows x 31 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "liberia = pd.concat(dfs).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Variable</th>\n",
       "      <th>Case Fatality Rate (CFR) - \n",
       " Confirmed &amp; Probable Cases</th>\n",
       "      <th>Case Fatality Rate (CFR) - Confirmed &amp; Probable Cases</th>\n",
       "      <th>Contacts lost to follow-up</th>\n",
       "      <th>Contacts seen</th>\n",
       "      <th>Contacts who completed 21 day \n",
       " follow-up</th>\n",
       "      <th>Contacts who completed 21 day follow-up</th>\n",
       "      <th>Cumulative (confirmed + probable + suspected)</th>\n",
       "      <th>Cumulative (confirmed + probable + suspects)</th>\n",
       "      <th>Cumulative CFR</th>\n",
       "      <th>Cumulative admission/isolation</th>\n",
       "      <th>...</th>\n",
       "      <th>Total death/s in confirmed,  probable, suspected cases</th>\n",
       "      <th>Total death/s in confirmed, probable, suspected cases</th>\n",
       "      <th>Total death/s in probable cases</th>\n",
       "      <th>Total death/s in suspected cases</th>\n",
       "      <th>Total discharges</th>\n",
       "      <th>Total no. currently in Treatment \n",
       " Units</th>\n",
       "      <th>Total no. currently in Treatment Units</th>\n",
       "      <th>Total probable cases</th>\n",
       "      <th>Total specimens tested</th>\n",
       "      <th>Total suspected cases</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-06-16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-22</th>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>224.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-25</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-28</th>\n",
       "      <td>NaN</td>\n",
       "      <td>61.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>407.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>49.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-06-29</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>49.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-07-01</th>\n",
       "      <td>NaN</td>\n",
       "      <td>70.1</td>\n",
       "      <td>15.0</td>\n",
       "      <td>422.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-07-02</th>\n",
       "      <td>NaN</td>\n",
       "      <td>74.3</td>\n",
       "      <td>15.0</td>\n",
       "      <td>423.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>66.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-07-03</th>\n",
       "      <td>NaN</td>\n",
       "      <td>75.3</td>\n",
       "      <td>14.0</td>\n",
       "      <td>423.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Variable    Case Fatality Rate (CFR) - \\n Confirmed & Probable Cases  \\\n",
       "Date                                                                   \n",
       "2014-06-16                                                NaN          \n",
       "2014-06-17                                                NaN          \n",
       "2014-06-22                                                NaN          \n",
       "2014-06-24                                                NaN          \n",
       "2014-06-25                                                NaN          \n",
       "2014-06-28                                                NaN          \n",
       "2014-06-29                                                NaN          \n",
       "2014-07-01                                                NaN          \n",
       "2014-07-02                                                NaN          \n",
       "2014-07-03                                                NaN          \n",
       "\n",
       "Variable    Case Fatality Rate (CFR) - Confirmed & Probable Cases  \\\n",
       "Date                                                                \n",
       "2014-06-16                                                NaN       \n",
       "2014-06-17                                                NaN       \n",
       "2014-06-22                                               25.0       \n",
       "2014-06-24                                                NaN       \n",
       "2014-06-25                                                NaN       \n",
       "2014-06-28                                               61.0       \n",
       "2014-06-29                                                NaN       \n",
       "2014-07-01                                               70.1       \n",
       "2014-07-02                                               74.3       \n",
       "2014-07-03                                               75.3       \n",
       "\n",
       "Variable    Contacts lost to follow-up  Contacts seen  \\\n",
       "Date                                                    \n",
       "2014-06-16                         NaN           95.0   \n",
       "2014-06-17                         0.0           90.0   \n",
       "2014-06-22                         0.0          232.0   \n",
       "2014-06-24                        15.0          224.0   \n",
       "2014-06-25                        15.0          112.0   \n",
       "2014-06-28                        15.0          407.0   \n",
       "2014-06-29                        15.0            0.0   \n",
       "2014-07-01                        15.0          422.0   \n",
       "2014-07-02                        15.0          423.0   \n",
       "2014-07-03                        14.0          423.0   \n",
       "\n",
       "Variable    Contacts who completed 21 day \\n follow-up  \\\n",
       "Date                                                     \n",
       "2014-06-16                                         NaN   \n",
       "2014-06-17                                         NaN   \n",
       "2014-06-22                                         NaN   \n",
       "2014-06-24                                         NaN   \n",
       "2014-06-25                                         NaN   \n",
       "2014-06-28                                         NaN   \n",
       "2014-06-29                                         NaN   \n",
       "2014-07-01                                         NaN   \n",
       "2014-07-02                                         NaN   \n",
       "2014-07-03                                         NaN   \n",
       "\n",
       "Variable    Contacts who completed 21 day follow-up  \\\n",
       "Date                                                  \n",
       "2014-06-16                                     11.0   \n",
       "2014-06-17                                     11.0   \n",
       "2014-06-22                                     17.0   \n",
       "2014-06-24                                     17.0   \n",
       "2014-06-25                                     17.0   \n",
       "2014-06-28                                     22.0   \n",
       "2014-06-29                                     22.0   \n",
       "2014-07-01                                     17.0   \n",
       "2014-07-02                                     40.0   \n",
       "2014-07-03                                     45.0   \n",
       "\n",
       "Variable    Cumulative (confirmed + probable + suspected)  \\\n",
       "Date                                                        \n",
       "2014-06-16                                            NaN   \n",
       "2014-06-17                                            NaN   \n",
       "2014-06-22                                            NaN   \n",
       "2014-06-24                                            NaN   \n",
       "2014-06-25                                            NaN   \n",
       "2014-06-28                                            NaN   \n",
       "2014-06-29                                            NaN   \n",
       "2014-07-01                                            NaN   \n",
       "2014-07-02                                            NaN   \n",
       "2014-07-03                                            NaN   \n",
       "\n",
       "Variable    Cumulative (confirmed + probable + suspects)  Cumulative CFR  \\\n",
       "Date                                                                       \n",
       "2014-06-16                                           NaN             NaN   \n",
       "2014-06-17                                           NaN             NaN   \n",
       "2014-06-22                                           NaN             NaN   \n",
       "2014-06-24                                           NaN             NaN   \n",
       "2014-06-25                                           NaN             NaN   \n",
       "2014-06-28                                           NaN             NaN   \n",
       "2014-06-29                                           NaN             NaN   \n",
       "2014-07-01                                           NaN             NaN   \n",
       "2014-07-02                                           NaN             NaN   \n",
       "2014-07-03                                           NaN             NaN   \n",
       "\n",
       "Variable    Cumulative admission/isolation          ...            \\\n",
       "Date                                                ...             \n",
       "2014-06-16                             5.0          ...             \n",
       "2014-06-17                            10.0          ...             \n",
       "2014-06-22                            15.0          ...             \n",
       "2014-06-24                            25.0          ...             \n",
       "2014-06-25                            25.0          ...             \n",
       "2014-06-28                            25.0          ...             \n",
       "2014-06-29                            25.0          ...             \n",
       "2014-07-01                            19.0          ...             \n",
       "2014-07-02                            18.0          ...             \n",
       "2014-07-03                            18.0          ...             \n",
       "\n",
       "Variable    Total death/s in confirmed,  probable, suspected cases  \\\n",
       "Date                                                                 \n",
       "2014-06-16                                                NaN        \n",
       "2014-06-17                                                NaN        \n",
       "2014-06-22                                                NaN        \n",
       "2014-06-24                                                NaN        \n",
       "2014-06-25                                                NaN        \n",
       "2014-06-28                                                NaN        \n",
       "2014-06-29                                                NaN        \n",
       "2014-07-01                                                NaN        \n",
       "2014-07-02                                                NaN        \n",
       "2014-07-03                                                NaN        \n",
       "\n",
       "Variable    Total death/s in confirmed, probable, suspected cases  \\\n",
       "Date                                                                \n",
       "2014-06-16                                               16.0       \n",
       "2014-06-17                                               16.0       \n",
       "2014-06-22                                               25.0       \n",
       "2014-06-24                                               32.0       \n",
       "2014-06-25                                               37.0       \n",
       "2014-06-28                                               49.0       \n",
       "2014-06-29                                               49.0       \n",
       "2014-07-01                                               61.0       \n",
       "2014-07-02                                               66.0       \n",
       "2014-07-03                                               70.0       \n",
       "\n",
       "Variable    Total death/s in probable cases  Total death/s in suspected cases  \\\n",
       "Date                                                                            \n",
       "2014-06-16                              6.0                               2.0   \n",
       "2014-06-17                              6.0                               2.0   \n",
       "2014-06-22                              8.0                               1.0   \n",
       "2014-06-24                              8.0                               6.0   \n",
       "2014-06-25                              8.0                               9.0   \n",
       "2014-06-28                             10.0                              13.0   \n",
       "2014-06-29                             10.0                              13.0   \n",
       "2014-07-01                             17.0                              14.0   \n",
       "2014-07-02                             20.0                              14.0   \n",
       "2014-07-03                             22.0                              15.0   \n",
       "\n",
       "Variable    Total discharges  Total no. currently in Treatment \\n Units  \\\n",
       "Date                                                                      \n",
       "2014-06-16               3.0                                        NaN   \n",
       "2014-06-17               0.0                                        NaN   \n",
       "2014-06-22               0.0                                        NaN   \n",
       "2014-06-24               0.0                                        NaN   \n",
       "2014-06-25               0.0                                        NaN   \n",
       "2014-06-28               3.0                                        NaN   \n",
       "2014-06-29               1.0                                        NaN   \n",
       "2014-07-01               3.0                                        NaN   \n",
       "2014-07-02               0.0                                        NaN   \n",
       "2014-07-03               0.0                                        NaN   \n",
       "\n",
       "Variable    Total no. currently in Treatment Units  Total probable cases  \\\n",
       "Date                                                                       \n",
       "2014-06-16                                     5.0                   6.0   \n",
       "2014-06-17                                    10.0                   6.0   \n",
       "2014-06-22                                    15.0                   8.0   \n",
       "2014-06-24                                    25.0                   8.0   \n",
       "2014-06-25                                    25.0                   9.0   \n",
       "2014-06-28                                    25.0                  18.0   \n",
       "2014-06-29                                    25.0                  18.0   \n",
       "2014-07-01                                    19.0                  19.0   \n",
       "2014-07-02                                    18.0                  22.0   \n",
       "2014-07-03                                    18.0                  24.0   \n",
       "\n",
       "Variable    Total specimens tested  Total suspected cases  \n",
       "Date                                                       \n",
       "2014-06-16                    28.0                    4.0  \n",
       "2014-06-17                    34.0                    8.0  \n",
       "2014-06-22                    49.0                    6.0  \n",
       "2014-06-24                    63.0                   13.0  \n",
       "2014-06-25                     4.0                   17.0  \n",
       "2014-06-28                     3.0                   29.0  \n",
       "2014-06-29                     4.0                   29.0  \n",
       "2014-07-01                     4.0                   33.0  \n",
       "2014-07-02                     0.0                   34.0  \n",
       "2014-07-03                     2.0                   35.0  \n",
       "\n",
       "[10 rows x 45 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "liberia.head(n=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sierra Leone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = {}\n",
    "tables_with_duplicates = []\n",
    "\n",
    "for file, path in ebola_csv_iter('sl_data'):\n",
    "    df = pd.read_csv(path,\n",
    "                     parse_dates=['date'],\n",
    "                     usecols=['date', 'variable', 'National'])\n",
    "    df.rename(columns={'date': 'Date', 'variable': 'Variable'}, inplace=True)\n",
    "    df.Variable = df.Variable.astype('category')\n",
    "    if len(df.Variable.unique()) != len(df.Variable):\n",
    "        tables_with_duplicates.append(file)\n",
    "    dfs[file] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['2014-11-10-v167.csv']"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tables_with_duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Date</th>\n",
       "      <th>Variable</th>\n",
       "      <th>National</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>2014-11-10</td>\n",
       "      <td>pending</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2014-11-10</td>\n",
       "      <td>positive_corpse</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>2014-11-10</td>\n",
       "      <td>negative_corpse</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>2014-11-10</td>\n",
       "      <td>pending</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>2014-11-10</td>\n",
       "      <td>positive_corpse</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>2014-11-10</td>\n",
       "      <td>negative_corpse</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Date         Variable  National\n",
       "34 2014-11-10          pending      32.0\n",
       "35 2014-11-10  positive_corpse      37.0\n",
       "36 2014-11-10  negative_corpse      91.0\n",
       "37 2014-11-10          pending       NaN\n",
       "38 2014-11-10  positive_corpse       NaN\n",
       "39 2014-11-10  negative_corpse       NaN"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with_duplicates = dfs[tables_with_duplicates[0]]\n",
    "with_duplicates[with_duplicates.duplicated(keep=False, subset='Variable')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This time we decide to drop `NaN` values in favor of present data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with_duplicates.drop_duplicates(subset='Variable', keep='first', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dfs = [df.pivot(index='Date', columns='Variable', values='National') for df in dfs.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Variable</th>\n",
       "      <th>cfr</th>\n",
       "      <th>contacts_followed</th>\n",
       "      <th>contacts_healthy</th>\n",
       "      <th>contacts_ill</th>\n",
       "      <th>contacts_not_seen</th>\n",
       "      <th>cum_completed_contacts</th>\n",
       "      <th>cum_confirmed</th>\n",
       "      <th>cum_contacts</th>\n",
       "      <th>cum_noncase</th>\n",
       "      <th>cum_probable</th>\n",
       "      <th>...</th>\n",
       "      <th>new_positive</th>\n",
       "      <th>new_probable</th>\n",
       "      <th>new_samples</th>\n",
       "      <th>new_suspected</th>\n",
       "      <th>pending</th>\n",
       "      <th>percent_seen</th>\n",
       "      <th>population</th>\n",
       "      <th>positive_corpse</th>\n",
       "      <th>repeat_samples</th>\n",
       "      <th>total_lab_samples</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-10-25</th>\n",
       "      <td>29.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3560.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3241.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>...</td>\n",
       "      <td>70.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6348350.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>240.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Variable     cfr  contacts_followed  contacts_healthy  contacts_ill  \\\n",
       "Date                                                                  \n",
       "2014-10-25  29.0                NaN               NaN           NaN   \n",
       "\n",
       "Variable    contacts_not_seen  cum_completed_contacts  cum_confirmed  \\\n",
       "Date                                                                   \n",
       "2014-10-25                NaN                     NaN         3560.0   \n",
       "\n",
       "Variable    cum_contacts  cum_noncase  cum_probable        ...          \\\n",
       "Date                                                       ...           \n",
       "2014-10-25           NaN       3241.0          82.0        ...           \n",
       "\n",
       "Variable    new_positive  new_probable  new_samples  new_suspected  pending  \\\n",
       "Date                                                                          \n",
       "2014-10-25          70.0           9.0        201.0            8.0      NaN   \n",
       "\n",
       "Variable    percent_seen  population  positive_corpse  repeat_samples  \\\n",
       "Date                                                                    \n",
       "2014-10-25           NaN   6348350.0             33.0            39.0   \n",
       "\n",
       "Variable    total_lab_samples  \n",
       "Date                           \n",
       "2014-10-25              240.0  \n",
       "\n",
       "[1 rows x 37 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sl = pd.concat(dfs).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Variable</th>\n",
       "      <th>cfr</th>\n",
       "      <th>contacts_followed</th>\n",
       "      <th>contacts_healthy</th>\n",
       "      <th>contacts_ill</th>\n",
       "      <th>contacts_not_seen</th>\n",
       "      <th>cum_completed_contacts</th>\n",
       "      <th>cum_confirmed</th>\n",
       "      <th>cum_contacts</th>\n",
       "      <th>cum_noncase</th>\n",
       "      <th>cum_probable</th>\n",
       "      <th>...</th>\n",
       "      <th>new_positive</th>\n",
       "      <th>new_probable</th>\n",
       "      <th>new_samples</th>\n",
       "      <th>new_suspected</th>\n",
       "      <th>pending</th>\n",
       "      <th>percent_seen</th>\n",
       "      <th>population</th>\n",
       "      <th>positive_corpse</th>\n",
       "      <th>repeat_samples</th>\n",
       "      <th>total_lab_samples</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-08-12</th>\n",
       "      <td>36.8</td>\n",
       "      <td>2255</td>\n",
       "      <td>1937</td>\n",
       "      <td>14</td>\n",
       "      <td>131</td>\n",
       "      <td>901</td>\n",
       "      <td>717</td>\n",
       "      <td>2838</td>\n",
       "      <td>657</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87%</td>\n",
       "      <td>6348350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-13</th>\n",
       "      <td>37.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1956</td>\n",
       "      <td>14</td>\n",
       "      <td>131</td>\n",
       "      <td>NaN</td>\n",
       "      <td>733</td>\n",
       "      <td>2099</td>\n",
       "      <td>680</td>\n",
       "      <td>38</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>94%</td>\n",
       "      <td>6,348,350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-14</th>\n",
       "      <td>37.5</td>\n",
       "      <td>2268</td>\n",
       "      <td>2135</td>\n",
       "      <td>7</td>\n",
       "      <td>117</td>\n",
       "      <td>1852</td>\n",
       "      <td>747</td>\n",
       "      <td>4173</td>\n",
       "      <td>708</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>94%</td>\n",
       "      <td>6,348,350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-15</th>\n",
       "      <td>37.9</td>\n",
       "      <td>2212</td>\n",
       "      <td>2137</td>\n",
       "      <td>7</td>\n",
       "      <td>95</td>\n",
       "      <td>1927</td>\n",
       "      <td>757</td>\n",
       "      <td>4280</td>\n",
       "      <td>726</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97%</td>\n",
       "      <td>6,348,350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-16</th>\n",
       "      <td>38.3</td>\n",
       "      <td>2323</td>\n",
       "      <td>2265</td>\n",
       "      <td>6</td>\n",
       "      <td>79</td>\n",
       "      <td>2021</td>\n",
       "      <td>775</td>\n",
       "      <td>4361</td>\n",
       "      <td>741</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98%</td>\n",
       "      <td>6,348,350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-17</th>\n",
       "      <td>39.2</td>\n",
       "      <td>2428</td>\n",
       "      <td>2311</td>\n",
       "      <td>8</td>\n",
       "      <td>97</td>\n",
       "      <td>2135</td>\n",
       "      <td>778</td>\n",
       "      <td>4451</td>\n",
       "      <td>758</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96%</td>\n",
       "      <td>6,348,350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-18</th>\n",
       "      <td>39.8</td>\n",
       "      <td>2510</td>\n",
       "      <td>2420</td>\n",
       "      <td>16</td>\n",
       "      <td>66</td>\n",
       "      <td>2243</td>\n",
       "      <td>783</td>\n",
       "      <td>4581</td>\n",
       "      <td>758</td>\n",
       "      <td>52</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97%</td>\n",
       "      <td>6,348,350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-19</th>\n",
       "      <td>39.8</td>\n",
       "      <td>2434</td>\n",
       "      <td>2347</td>\n",
       "      <td>14</td>\n",
       "      <td>71</td>\n",
       "      <td>2313</td>\n",
       "      <td>804</td>\n",
       "      <td>4686</td>\n",
       "      <td>802</td>\n",
       "      <td>40</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97%</td>\n",
       "      <td>6348350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-20</th>\n",
       "      <td>39.6</td>\n",
       "      <td>2423</td>\n",
       "      <td>2227</td>\n",
       "      <td>11</td>\n",
       "      <td>172</td>\n",
       "      <td>2384</td>\n",
       "      <td>813</td>\n",
       "      <td>4804</td>\n",
       "      <td>817</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>92%</td>\n",
       "      <td>6348350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-21</th>\n",
       "      <td>40</td>\n",
       "      <td>3026</td>\n",
       "      <td>2273</td>\n",
       "      <td>7</td>\n",
       "      <td>81</td>\n",
       "      <td>1901</td>\n",
       "      <td>823</td>\n",
       "      <td>4896</td>\n",
       "      <td>825</td>\n",
       "      <td>38</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75%</td>\n",
       "      <td>6348350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Variable     cfr contacts_followed contacts_healthy contacts_ill  \\\n",
       "Date                                                               \n",
       "2014-08-12  36.8              2255             1937           14   \n",
       "2014-08-13  37.2               NaN             1956           14   \n",
       "2014-08-14  37.5              2268             2135            7   \n",
       "2014-08-15  37.9              2212             2137            7   \n",
       "2014-08-16  38.3              2323             2265            6   \n",
       "2014-08-17  39.2              2428             2311            8   \n",
       "2014-08-18  39.8              2510             2420           16   \n",
       "2014-08-19  39.8              2434             2347           14   \n",
       "2014-08-20  39.6              2423             2227           11   \n",
       "2014-08-21    40              3026             2273            7   \n",
       "\n",
       "Variable   contacts_not_seen cum_completed_contacts cum_confirmed  \\\n",
       "Date                                                                \n",
       "2014-08-12               131                    901           717   \n",
       "2014-08-13               131                    NaN           733   \n",
       "2014-08-14               117                   1852           747   \n",
       "2014-08-15                95                   1927           757   \n",
       "2014-08-16                79                   2021           775   \n",
       "2014-08-17                97                   2135           778   \n",
       "2014-08-18                66                   2243           783   \n",
       "2014-08-19                71                   2313           804   \n",
       "2014-08-20               172                   2384           813   \n",
       "2014-08-21                81                   1901           823   \n",
       "\n",
       "Variable   cum_contacts cum_noncase cum_probable        ...         \\\n",
       "Date                                                    ...          \n",
       "2014-08-12         2838         657           37        ...          \n",
       "2014-08-13         2099         680           38        ...          \n",
       "2014-08-14         4173         708           39        ...          \n",
       "2014-08-15         4280         726           37        ...          \n",
       "2014-08-16         4361         741           34        ...          \n",
       "2014-08-17         4451         758           37        ...          \n",
       "2014-08-18         4581         758           52        ...          \n",
       "2014-08-19         4686         802           40        ...          \n",
       "2014-08-20         4804         817           37        ...          \n",
       "2014-08-21         4896         825           38        ...          \n",
       "\n",
       "Variable   new_positive new_probable new_samples new_suspected pending  \\\n",
       "Date                                                                     \n",
       "2014-08-12          NaN            1         NaN            10     NaN   \n",
       "2014-08-13          NaN            1         NaN             3     NaN   \n",
       "2014-08-14          NaN            2         NaN             0     NaN   \n",
       "2014-08-15          NaN            1         NaN             6     NaN   \n",
       "2014-08-16          NaN            0         NaN             3     NaN   \n",
       "2014-08-17          NaN            0         NaN             1     NaN   \n",
       "2014-08-18          NaN           15         NaN            40     NaN   \n",
       "2014-08-19          NaN            3         NaN            16     NaN   \n",
       "2014-08-20          NaN            0         NaN             1     NaN   \n",
       "2014-08-21          NaN            0         NaN             0     NaN   \n",
       "\n",
       "Variable   percent_seen population positive_corpse repeat_samples  \\\n",
       "Date                                                                \n",
       "2014-08-12          87%    6348350             NaN            NaN   \n",
       "2014-08-13          94%  6,348,350             NaN            NaN   \n",
       "2014-08-14          94%  6,348,350             NaN            NaN   \n",
       "2014-08-15          97%  6,348,350             NaN            NaN   \n",
       "2014-08-16          98%  6,348,350             NaN            NaN   \n",
       "2014-08-17          96%  6,348,350             NaN            NaN   \n",
       "2014-08-18          97%  6,348,350             NaN            NaN   \n",
       "2014-08-19          97%    6348350             NaN            NaN   \n",
       "2014-08-20          92%    6348350             NaN            NaN   \n",
       "2014-08-21          75%    6348350             NaN            NaN   \n",
       "\n",
       "Variable   total_lab_samples  \n",
       "Date                          \n",
       "2014-08-12               NaN  \n",
       "2014-08-13               NaN  \n",
       "2014-08-14               NaN  \n",
       "2014-08-15               NaN  \n",
       "2014-08-16               NaN  \n",
       "2014-08-17               NaN  \n",
       "2014-08-18               NaN  \n",
       "2014-08-19               NaN  \n",
       "2014-08-20               NaN  \n",
       "2014-08-21               NaN  \n",
       "\n",
       "[10 rows x 37 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sl.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "del dfs\n",
    "del tables_with_duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combining the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CategoricalIndex(['Cumulative (confirmed + probable + suspects)',\n",
       "                  'Fatality rate for confirmed and probables',\n",
       "                  'New admits to CTE so far', 'New cases of confirmed',\n",
       "                  'New cases of confirmed among health workers',\n",
       "                  'New cases of probables', 'New cases of suspects',\n",
       "                  'New contacts registered so far', 'New deaths registered',\n",
       "                  'New deaths registered among health workers',\n",
       "                  'New deaths registered today',\n",
       "                  'New deaths registered today (confirmed)',\n",
       "                  'New deaths registered today (probables)',\n",
       "                  'New deaths registered today (suspects)',\n",
       "                  'Number of confirmed cases among health workers',\n",
       "                  'Number of contacts followed today',\n",
       "                  'Number of contacts followed yesterday',\n",
       "                  'Number of contacts lost to follow up',\n",
       "                  'Number of contacts out of the track 21 days',\n",
       "                  'Number of contacts out of track',\n",
       "                  'Number of contacts to follow today',\n",
       "                  'Number of death of confirmed cases among health workers',\n",
       "                  'Number of deaths of confirmed cases among health workers',\n",
       "                  'Number of deaths of probables cases among health workers',\n",
       "                  'Number of female confirmed cases',\n",
       "                  'Number of female probables cases',\n",
       "                  'Number of female suspects cases',\n",
       "                  'Number of male confirmed cases',\n",
       "                  'Number of male probables cases',\n",
       "                  'Number of male suspects cases', 'Number of patients tested',\n",
       "                  'Number of probables cases among health workers',\n",
       "                  'Number of samples collected',\n",
       "                  'Number of samples collected today',\n",
       "                  'Number of samples under test',\n",
       "                  'Number of suspects cases among health workers',\n",
       "                  'Total PEC center today',\n",
       "                  'Total PEC center today (confirmed)',\n",
       "                  'Total PEC center today (probables)',\n",
       "                  'Total PEC center today (suspects)',\n",
       "                  'Total case of confirmed among health workers',\n",
       "                  'Total cases of confirmed', 'Total cases of probables',\n",
       "                  'Total cases of suspects',\n",
       "                  'Total contacts registered from start date',\n",
       "                  'Total deaths (confirmed + probables + suspects)',\n",
       "                  'Total deaths of confirmed', 'Total deaths of probables',\n",
       "                  'Total deaths of suspects',\n",
       "                  'Total deaths registered among health workers',\n",
       "                  'Total new cases registered so far',\n",
       "                  'Total number of admissions to CTE',\n",
       "                  'Total number of exits from CTE',\n",
       "                  'Total number of female cases',\n",
       "                  'Total number of hospitalized cases in CTE',\n",
       "                  'Total number of male cases',\n",
       "                  'Total of cured in confirmed cases in CTE',\n",
       "                  'Total of deaths in confirmed cases in CTE',\n",
       "                  'Total samples tested', 'Total suspected non-class cases'],\n",
       "                 categories=['Cumulative (confirmed + probable + suspects)', 'Fatality rate for confirmed and probables', 'New admits to CTE so far', 'New cases of confirmed', 'New cases of confirmed among health workers', 'New cases of probables', 'New cases of suspects', 'New contacts registered so far', ...], ordered=False, name='Variable', dtype='category')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guinea.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "liberia.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sl.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup for location of the dataset of the task 2\n",
    "MICROBIOME_FOLDER = DATA_FOLDER + \"/microbiome\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your answer here\n",
    "# TODO: Should the data from the spreadsheets be merged in different columns instead of adding indexes as we are doing now\n",
    "#        Should at least the tissue and stool from the same group of patient be merged?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Basic analysis of the file formatting:\n",
    "\n",
    "***\n",
    "**For the files MIDn.xls with n in [1,9]**\n",
    "\n",
    "**column 1:**\n",
    "\n",
    "We see that the first column in the files contain the scientific classification of the microbiomes\n",
    "Altough it could be kept as a single string, it would have more meaning if splitted\n",
    "The scientific classification contains the following subdivisions (https://en.wikipedia.org/wiki/Taxonomic_rank): \n",
    "\n",
    "       Domain, Kingdom, Phylum, Class, Order, Family, Genus, Species\n",
    "       \n",
    " With this classification, we have a problem: only 6 strings are given in the data set, whilst we have 8 potential divisions in the classification.\n",
    " \n",
    " After analysis, it seems that Kingdom is not used with bacterias (https://en.wikipedia.org/wiki/Bacteria)\n",
    " and always the same as the Phylum in Archea(https://en.wikipedia.org/wiki/Archaea). Therefore, **we will not use Kingdom as a division**.\n",
    " \n",
    " Also, Species only has a sense in the Eucaryote Domain, for which we have no data, so **we will not use Species as a division**.\n",
    " \n",
    " In addition, after playing with the data, we found cases where the Family can be named: \"Incertae Sedis\", which would be classified as the family and genus in our classification. Therefore, we need to check for those cases and re-concatenate the strings to return a proper list in every cases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lets keep those names in a list for further use\n",
    "scientific_classification = [\"domain\", \"phylum\", \"class\", \"order\", \"family\", \"genus\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Column 2:**\n",
    "\n",
    "We see that the second column contains an integer value. We can suppose that this value is the number of samples\n",
    "containing the genus defined by the first column. There's no title to the column, so we don't know yet where\n",
    "those values came from.\n",
    "\n",
    "***\n",
    "\n",
    "**For the file \"Metadata.xls\"**\n",
    "\n",
    "**Column 1:**\n",
    "\n",
    "Titled **\"BARCODE\"**, give xls file identifier for which the two other columns give more information\n",
    "\n",
    "**Column 2:**\n",
    "\n",
    "Titled **\"GROUP\"**, gives the groups from which each dataset has been sampled. Two informations are contained in those groups, the first is given by \"NEC\", \"Control\" or \"EXTRACTION CONTROL\" and the second is the numbering of the group (which is likely a phase of test), either \"1\" or \"2\". We will split those informations in two column, because while using the dataset, we might want to combine all the \"NEC\" patients or all the patient for a specific phase.\n",
    "\n",
    "**Column 3:**\n",
    "\n",
    "Titled **\"SAMPLE\"**, gives the type of sample that was taken, either tissue, stool or NA. Each group had both types of samples taken. \n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Desired formating of the data after analysis\n",
    "\n",
    "The simple analysis above tells us what are the columns that we will want in our Data Frame\n",
    "\n",
    "1. **6** column for classification, 1 for each classifier in the \"scientific_classification\" list. This will caracterise each microbiome individually.\n",
    "\n",
    "2. **1** column will contain the value associated with each microbiome measurement.\n",
    "\n",
    "3. **2** columns describing the group of the sample from the metadata. The first column will be called **Group Type** and will contain the \"NEC\", \"Control\" or \"EXTRACTION CONTROL\" value. The second column will be called **Group Phase** and will contain either \"1\", \"2\" or \"unknow\", \"unknow\" will be used for the \"EXCTRACTON CONTROL\" group.\n",
    "\n",
    "4. **1** column describing the type of sample taken from the metadata. This column will be called **Sample** This will be either \"tissue\", \"stool\" or \"unknow\", \"unknow\" in the case of the EXTRACTION CONTROL group.\n",
    "\n",
    "For a total of **10** columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['domain',\n",
       " 'phylum',\n",
       " 'class',\n",
       " 'order',\n",
       " 'family',\n",
       " 'genus',\n",
       " 'value',\n",
       " 'group_type',\n",
       " 'group_phase',\n",
       " 'sample']"
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_col_list = scientific_classification + [\"value\", \"group_type\", \"group_phase\", \"sample\"]\n",
    "df_col_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### We want to extract the metadata from the metadata.xls file\n",
    "\n",
    "The metadata is needed before we start extracting the data from the MB files to create the DataFrame with all the desired columns.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The actual name of the excel sheet is \"Sheet1\" and not \"Sheet 1\" as in the other files.\n",
    "metadata_raw = pd.read_excel(MICROBIOME_FOLDER+\"/metadata.xls\", sheetname='Sheet1', header=0)\n",
    "metadata_raw.columns = metadata_raw.columns.str.lower()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Extract group phase and group type\n",
    "group = metadata_raw[\"group\"]\n",
    "group_type = []\n",
    "group_phase = []\n",
    "for index in range(len(group)):                # Index progression must be upward!\n",
    "    # A special case for the extraction control, we don't want to split it\n",
    "    if group[index] == \"EXTRACTION CONTROL\":\n",
    "        group_type.append(group[index])\n",
    "        group_phase.append(\"unknow\")\n",
    "    else:\n",
    "        splitted_group = group[index].split()\n",
    "        group_type.append(splitted_group[0])\n",
    "        group_phase.append(splitted_group[1])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>sample</th>\n",
       "      <th>group_type</th>\n",
       "      <th>group_phase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MID1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MID2</td>\n",
       "      <td>tissue</td>\n",
       "      <td>NEC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MID3</td>\n",
       "      <td>tissue</td>\n",
       "      <td>Control</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MID4</td>\n",
       "      <td>tissue</td>\n",
       "      <td>NEC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MID5</td>\n",
       "      <td>tissue</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MID6</td>\n",
       "      <td>stool</td>\n",
       "      <td>NEC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MID7</td>\n",
       "      <td>stool</td>\n",
       "      <td>Control</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MID8</td>\n",
       "      <td>stool</td>\n",
       "      <td>NEC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MID9</td>\n",
       "      <td>stool</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  barcode  sample          group_type group_phase\n",
       "0    MID1     NaN  EXTRACTION CONTROL      unknow\n",
       "1    MID2  tissue                 NEC           1\n",
       "2    MID3  tissue             Control           1\n",
       "3    MID4  tissue                 NEC           2\n",
       "4    MID5  tissue             Control           2\n",
       "5    MID6   stool                 NEC           1\n",
       "6    MID7   stool             Control           1\n",
       "7    MID8   stool                 NEC           2\n",
       "8    MID9   stool             Control           2"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_raw['group_type'] = pd.Series(group_type)\n",
    "metadata_raw['group_phase'] = pd.Series(group_phase)\n",
    "metadata = metadata_raw.drop(\"group\",  axis=1)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>sample</th>\n",
       "      <th>group_type</th>\n",
       "      <th>group_phase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MID1</td>\n",
       "      <td>unknow</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MID2</td>\n",
       "      <td>tissue</td>\n",
       "      <td>NEC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MID3</td>\n",
       "      <td>tissue</td>\n",
       "      <td>Control</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MID4</td>\n",
       "      <td>tissue</td>\n",
       "      <td>NEC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MID5</td>\n",
       "      <td>tissue</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MID6</td>\n",
       "      <td>stool</td>\n",
       "      <td>NEC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MID7</td>\n",
       "      <td>stool</td>\n",
       "      <td>Control</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MID8</td>\n",
       "      <td>stool</td>\n",
       "      <td>NEC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MID9</td>\n",
       "      <td>stool</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  barcode  sample          group_type group_phase\n",
       "0    MID1  unknow  EXTRACTION CONTROL      unknow\n",
       "1    MID2  tissue                 NEC           1\n",
       "2    MID3  tissue             Control           1\n",
       "3    MID4  tissue                 NEC           2\n",
       "4    MID5  tissue             Control           2\n",
       "5    MID6   stool                 NEC           1\n",
       "6    MID7   stool             Control           1\n",
       "7    MID8   stool                 NEC           2\n",
       "8    MID9   stool             Control           2"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## make sure SAMPLE doesn't contain NaN\n",
    "metadata[\"sample\"].fillna(value=\"unknow\", inplace=True)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The format of the metadata is now as we want it"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### We now want to extract the data from the datasheets\n",
    "\n",
    "** We create a function to extract the classifications **\n",
    "\n",
    "We want to extract the classifications from the first column of the datasheets. We will need to do string parsing. Regex are a great tool to remove the unwanted characters from the format. The only unwanted characters are the double quotes \". Also, as said before, we need to manage the case where we have the name \"Incertae Sedis\" as a Family (4th position)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# TODO: this could be made more general, for every potential cases where there's more that 6 elements\n",
    "# Next function replace \" in the classifier_string, split the string to have an indexable list and return it\n",
    "def get_classifiers(classifier_string):\n",
    "    classified_list = re.subn(\"\\\"\", \"\", classifier_string)[0].split()\n",
    "    \n",
    "    # Special case management\n",
    "    if len(classified_list) > 6:\n",
    "        # We join the first extra location with the family\n",
    "        classified_list[4] = classified_list[4] + \"_\" + classified_list[5]\n",
    "        del classified_list[5]\n",
    "        # Use a loop to allow for harbitrary number of words in the genus and for safety\n",
    "        while len(classified_list) > 6:\n",
    "            classified_list[5] += \"_\" + classified_list[6]\n",
    "            del classified_list[6]\n",
    "        \n",
    "    return classified_list\n",
    "    # Check if the 4th value is \"Incertae\"\n",
    "#    if \"Incertae\" in classified_list:\n",
    "#        classified_list[4] = classified_list[4] + \"_\" + classified_list[5]\n",
    "#        if classified_list[7]:\n",
    "#            classified_list[5] = classified_list[6] + \"_\" + classified_list[7]\n",
    "#            classified_list.pop()\n",
    "#            classified_list.pop()\n",
    "#        else:\n",
    "#            classified_list[5] = classified_list[6]\n",
    "#            classified_list.pop()\n",
    "#    return classified_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** We can now attempt to create the desired dataframe **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Desulfurococcales</td>\n",
       "      <td>Desulfurococcaceae</td>\n",
       "      <td>Ignisphaera</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Desulfurococcales</td>\n",
       "      <td>Pyrodictiaceae</td>\n",
       "      <td>Pyrolobus</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Sulfolobales</td>\n",
       "      <td>Sulfolobaceae</td>\n",
       "      <td>Stygiolobus</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Thermoproteales</td>\n",
       "      <td>Thermofilaceae</td>\n",
       "      <td>Thermofilum</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomicrobia</td>\n",
       "      <td>Methanocellales</td>\n",
       "      <td>Methanocellaceae</td>\n",
       "      <td>Methanocella</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomicrobia</td>\n",
       "      <td>Methanosarcinales</td>\n",
       "      <td>Methanosarcinaceae</td>\n",
       "      <td>Methanimicrococcus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomicrobia</td>\n",
       "      <td>Methanosarcinales</td>\n",
       "      <td>Methermicoccaceae</td>\n",
       "      <td>Methermicoccus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Archaeoglobales</td>\n",
       "      <td>Archaeoglobaceae</td>\n",
       "      <td>Ferroglobus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Archaeoglobales</td>\n",
       "      <td>Archaeoglobaceae</td>\n",
       "      <td>Geoglobus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Haloplanus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Haloquadratum</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Halosimplex</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Natronobacterium</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Natronomonas</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanococci</td>\n",
       "      <td>Methanococcales</td>\n",
       "      <td>Methanocaldococcaceae</td>\n",
       "      <td>Methanotorris</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanopyri</td>\n",
       "      <td>Methanopyrales</td>\n",
       "      <td>Methanopyraceae</td>\n",
       "      <td>Methanopyrus</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Thermoplasmatales</td>\n",
       "      <td>Thermoplasmatales_incertae_sedis</td>\n",
       "      <td>Thermogymnomonas</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Acidothermaceae</td>\n",
       "      <td>Acidothermus</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Beutenbergiaceae</td>\n",
       "      <td>Salana</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Bogoriellaceae</td>\n",
       "      <td>Bogoriella</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Cellulomonadaceae</td>\n",
       "      <td>Tropheryma</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Corynebacteriaceae</td>\n",
       "      <td>Turicella</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Cryptosporangiaceae</td>\n",
       "      <td>Cryptosporangium</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Dermabacteraceae</td>\n",
       "      <td>Dermabacter</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Dermabacteraceae</td>\n",
       "      <td>Devriesea</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Dermatophilaceae</td>\n",
       "      <td>Kineosphaera</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Frankineae_incertae_sedis</td>\n",
       "      <td>Fodinicola</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Geodermatophilaceae</td>\n",
       "      <td>Blastococcus</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Geodermatophilaceae</td>\n",
       "      <td>Geodermatophilus</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Intrasporangiaceae</td>\n",
       "      <td>Kribbia</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oceanospirillaceae</td>\n",
       "      <td>Neptuniibacter</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oleiphilaceae</td>\n",
       "      <td>Oleiphilus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Acinetobacter</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Alkanindiges</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Enhydrobacter</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Azomonas</td>\n",
       "      <td>396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Azorhizophilus</td>\n",
       "      <td>414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>3732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Piscirickettsiaceae</td>\n",
       "      <td>Cycloclasticus</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Thiotrichaceae</td>\n",
       "      <td>Beggiatoa</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Thiotrichaceae</td>\n",
       "      <td>Leucothrix</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Thiotrichales_incertae_sedis</td>\n",
       "      <td>Fangia</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Sinobacteraceae</td>\n",
       "      <td>Alkanibacter</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Sinobacteraceae</td>\n",
       "      <td>Hydrocarboniphaga</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Sinobacteraceae</td>\n",
       "      <td>Sinobacter</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Sinobacteraceae</td>\n",
       "      <td>Steroidobacter</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Frateuria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Luteibacter</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>Spirochaetales</td>\n",
       "      <td>Spirochaetales_incertae_sedis</td>\n",
       "      <td>Exilispira</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Synergistetes</td>\n",
       "      <td>Synergistia</td>\n",
       "      <td>Synergistales</td>\n",
       "      <td>Synergistaceae</td>\n",
       "      <td>Aminiphilus</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Synergistetes</td>\n",
       "      <td>Synergistia</td>\n",
       "      <td>Synergistales</td>\n",
       "      <td>Synergistaceae</td>\n",
       "      <td>Aminomonas</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Synergistetes</td>\n",
       "      <td>Synergistia</td>\n",
       "      <td>Synergistales</td>\n",
       "      <td>Synergistaceae</td>\n",
       "      <td>Thermovirga</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>Mollicutes</td>\n",
       "      <td>Haloplasmatales</td>\n",
       "      <td>Haloplasmataceae</td>\n",
       "      <td>Haloplasma</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteriales</td>\n",
       "      <td>Thermodesulfobacteriaceae</td>\n",
       "      <td>Caldimicrobium</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogales</td>\n",
       "      <td>Thermotogaceae</td>\n",
       "      <td>Geotoga</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogales</td>\n",
       "      <td>Thermotogaceae</td>\n",
       "      <td>Kosmotoga</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Verrucomicrobia</td>\n",
       "      <td>Opitutae</td>\n",
       "      <td>Opitutales</td>\n",
       "      <td>Opitutaceae</td>\n",
       "      <td>Alterococcus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Bangiophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Chlorarachniophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>None</td>\n",
       "      <td>1388</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>272 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       domain                 phylum                  class  \\\n",
       "0     Archaea          Crenarchaeota           Thermoprotei   \n",
       "1     Archaea          Crenarchaeota           Thermoprotei   \n",
       "2     Archaea          Crenarchaeota           Thermoprotei   \n",
       "3     Archaea          Crenarchaeota           Thermoprotei   \n",
       "4     Archaea          Euryarchaeota        Methanomicrobia   \n",
       "5     Archaea          Euryarchaeota        Methanomicrobia   \n",
       "6     Archaea          Euryarchaeota        Methanomicrobia   \n",
       "7     Archaea          Euryarchaeota           Archaeoglobi   \n",
       "8     Archaea          Euryarchaeota           Archaeoglobi   \n",
       "9     Archaea          Euryarchaeota           Halobacteria   \n",
       "10    Archaea          Euryarchaeota           Halobacteria   \n",
       "11    Archaea          Euryarchaeota           Halobacteria   \n",
       "12    Archaea          Euryarchaeota           Halobacteria   \n",
       "13    Archaea          Euryarchaeota           Halobacteria   \n",
       "14    Archaea          Euryarchaeota           Methanococci   \n",
       "15    Archaea          Euryarchaeota            Methanopyri   \n",
       "16    Archaea          Euryarchaeota         Thermoplasmata   \n",
       "17   Bacteria         Actinobacteria         Actinobacteria   \n",
       "18   Bacteria         Actinobacteria         Actinobacteria   \n",
       "19   Bacteria         Actinobacteria         Actinobacteria   \n",
       "20   Bacteria         Actinobacteria         Actinobacteria   \n",
       "21   Bacteria         Actinobacteria         Actinobacteria   \n",
       "22   Bacteria         Actinobacteria         Actinobacteria   \n",
       "23   Bacteria         Actinobacteria         Actinobacteria   \n",
       "24   Bacteria         Actinobacteria         Actinobacteria   \n",
       "25   Bacteria         Actinobacteria         Actinobacteria   \n",
       "26   Bacteria         Actinobacteria         Actinobacteria   \n",
       "27   Bacteria         Actinobacteria         Actinobacteria   \n",
       "28   Bacteria         Actinobacteria         Actinobacteria   \n",
       "29   Bacteria         Actinobacteria         Actinobacteria   \n",
       "..        ...                    ...                    ...   \n",
       "242  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "243  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "244  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "245  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "246  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "247  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "248  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "249  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "250  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "251  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "252  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "253  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "254  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "255  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "256  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "257  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "258  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "259  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "260  Bacteria           Spirochaetes           Spirochaetes   \n",
       "261  Bacteria          Synergistetes            Synergistia   \n",
       "262  Bacteria          Synergistetes            Synergistia   \n",
       "263  Bacteria          Synergistetes            Synergistia   \n",
       "264  Bacteria            Tenericutes             Mollicutes   \n",
       "265  Bacteria  Thermodesulfobacteria  Thermodesulfobacteria   \n",
       "266  Bacteria            Thermotogae            Thermotogae   \n",
       "267  Bacteria            Thermotogae            Thermotogae   \n",
       "268  Bacteria        Verrucomicrobia               Opitutae   \n",
       "269  Bacteria          Cyanobacteria          Cyanobacteria   \n",
       "270  Bacteria          Cyanobacteria          Cyanobacteria   \n",
       "271  Bacteria          Cyanobacteria          Cyanobacteria   \n",
       "\n",
       "                        order                            family  \\\n",
       "0           Desulfurococcales                Desulfurococcaceae   \n",
       "1           Desulfurococcales                    Pyrodictiaceae   \n",
       "2                Sulfolobales                     Sulfolobaceae   \n",
       "3             Thermoproteales                    Thermofilaceae   \n",
       "4             Methanocellales                  Methanocellaceae   \n",
       "5           Methanosarcinales                Methanosarcinaceae   \n",
       "6           Methanosarcinales                 Methermicoccaceae   \n",
       "7             Archaeoglobales                  Archaeoglobaceae   \n",
       "8             Archaeoglobales                  Archaeoglobaceae   \n",
       "9             Halobacteriales                  Halobacteriaceae   \n",
       "10            Halobacteriales                  Halobacteriaceae   \n",
       "11            Halobacteriales                  Halobacteriaceae   \n",
       "12            Halobacteriales                  Halobacteriaceae   \n",
       "13            Halobacteriales                  Halobacteriaceae   \n",
       "14            Methanococcales             Methanocaldococcaceae   \n",
       "15             Methanopyrales                   Methanopyraceae   \n",
       "16          Thermoplasmatales  Thermoplasmatales_incertae_sedis   \n",
       "17            Actinomycetales                   Acidothermaceae   \n",
       "18            Actinomycetales                  Beutenbergiaceae   \n",
       "19            Actinomycetales                    Bogoriellaceae   \n",
       "20            Actinomycetales                 Cellulomonadaceae   \n",
       "21            Actinomycetales                Corynebacteriaceae   \n",
       "22            Actinomycetales               Cryptosporangiaceae   \n",
       "23            Actinomycetales                  Dermabacteraceae   \n",
       "24            Actinomycetales                  Dermabacteraceae   \n",
       "25            Actinomycetales                  Dermatophilaceae   \n",
       "26            Actinomycetales         Frankineae_incertae_sedis   \n",
       "27            Actinomycetales               Geodermatophilaceae   \n",
       "28            Actinomycetales               Geodermatophilaceae   \n",
       "29            Actinomycetales                Intrasporangiaceae   \n",
       "..                        ...                               ...   \n",
       "242         Oceanospirillales                Oceanospirillaceae   \n",
       "243         Oceanospirillales                     Oleiphilaceae   \n",
       "244           Pseudomonadales                     Moraxellaceae   \n",
       "245           Pseudomonadales                     Moraxellaceae   \n",
       "246           Pseudomonadales                     Moraxellaceae   \n",
       "247           Pseudomonadales                  Pseudomonadaceae   \n",
       "248           Pseudomonadales                  Pseudomonadaceae   \n",
       "249           Pseudomonadales                  Pseudomonadaceae   \n",
       "250             Thiotrichales               Piscirickettsiaceae   \n",
       "251             Thiotrichales                    Thiotrichaceae   \n",
       "252             Thiotrichales                    Thiotrichaceae   \n",
       "253             Thiotrichales      Thiotrichales_incertae_sedis   \n",
       "254           Xanthomonadales                   Sinobacteraceae   \n",
       "255           Xanthomonadales                   Sinobacteraceae   \n",
       "256           Xanthomonadales                   Sinobacteraceae   \n",
       "257           Xanthomonadales                   Sinobacteraceae   \n",
       "258           Xanthomonadales                  Xanthomonadaceae   \n",
       "259           Xanthomonadales                  Xanthomonadaceae   \n",
       "260            Spirochaetales     Spirochaetales_incertae_sedis   \n",
       "261             Synergistales                    Synergistaceae   \n",
       "262             Synergistales                    Synergistaceae   \n",
       "263             Synergistales                    Synergistaceae   \n",
       "264           Haloplasmatales                  Haloplasmataceae   \n",
       "265  Thermodesulfobacteriales         Thermodesulfobacteriaceae   \n",
       "266             Thermotogales                    Thermotogaceae   \n",
       "267             Thermotogales                    Thermotogaceae   \n",
       "268                Opitutales                       Opitutaceae   \n",
       "269               Chloroplast                     Bangiophyceae   \n",
       "270               Chloroplast              Chlorarachniophyceae   \n",
       "271               Chloroplast                      Streptophyta   \n",
       "\n",
       "                  genus  value  \n",
       "0           Ignisphaera      7  \n",
       "1             Pyrolobus      2  \n",
       "2           Stygiolobus      3  \n",
       "3           Thermofilum      3  \n",
       "4          Methanocella      7  \n",
       "5    Methanimicrococcus      1  \n",
       "6        Methermicoccus      1  \n",
       "7           Ferroglobus      1  \n",
       "8             Geoglobus      1  \n",
       "9            Haloplanus      1  \n",
       "10        Haloquadratum      4  \n",
       "11          Halosimplex      1  \n",
       "12     Natronobacterium      2  \n",
       "13         Natronomonas      4  \n",
       "14        Methanotorris      1  \n",
       "15         Methanopyrus     12  \n",
       "16     Thermogymnomonas      2  \n",
       "17         Acidothermus      2  \n",
       "18               Salana      1  \n",
       "19           Bogoriella      1  \n",
       "20           Tropheryma      1  \n",
       "21            Turicella      2  \n",
       "22     Cryptosporangium     53  \n",
       "23          Dermabacter      1  \n",
       "24            Devriesea      2  \n",
       "25         Kineosphaera      3  \n",
       "26           Fodinicola      1  \n",
       "27         Blastococcus     99  \n",
       "28     Geodermatophilus     17  \n",
       "29              Kribbia      2  \n",
       "..                  ...    ...  \n",
       "242      Neptuniibacter      1  \n",
       "243          Oleiphilus      1  \n",
       "244       Acinetobacter     11  \n",
       "245        Alkanindiges      9  \n",
       "246       Enhydrobacter      1  \n",
       "247            Azomonas    396  \n",
       "248      Azorhizophilus    414  \n",
       "249         Pseudomonas   3732  \n",
       "250      Cycloclasticus      4  \n",
       "251           Beggiatoa     22  \n",
       "252          Leucothrix      1  \n",
       "253              Fangia      4  \n",
       "254        Alkanibacter      5  \n",
       "255   Hydrocarboniphaga      6  \n",
       "256          Sinobacter      1  \n",
       "257      Steroidobacter      1  \n",
       "258           Frateuria      2  \n",
       "259         Luteibacter      2  \n",
       "260          Exilispira      1  \n",
       "261         Aminiphilus      2  \n",
       "262          Aminomonas      1  \n",
       "263         Thermovirga      1  \n",
       "264          Haloplasma      1  \n",
       "265      Caldimicrobium      1  \n",
       "266             Geotoga      7  \n",
       "267           Kosmotoga      9  \n",
       "268        Alterococcus      1  \n",
       "269                None      2  \n",
       "270                None     85  \n",
       "271                None   1388  \n",
       "\n",
       "[272 rows x 7 columns]"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# In this section, we test the analysis of a single datasheet\n",
    "\n",
    "test_data = pd.read_excel(MICROBIOME_FOLDER+\"/MID1.xls\", sheetname='Sheet 1', header=None)\n",
    "test_data.columns = [\"raw_classification\", \"value\"]\n",
    "classifier_df = pd.DataFrame(columns=df_col_list)\n",
    "classifier_array = []\n",
    "for row_data in test_data.itertuples():\n",
    "    classifier_array.append(get_classifiers(row_data.raw_classification))\n",
    "classifier_series = pd.DataFrame(classifier_array, columns=scientific_classification)\n",
    "classifier_series[\"value\"] = test_data[\"value\"]\n",
    "classifier_series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>value</th>\n",
       "      <th>group_type</th>\n",
       "      <th>group_phase</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Desulfurococcales</td>\n",
       "      <td>Desulfurococcaceae</td>\n",
       "      <td>Ignisphaera</td>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Desulfurococcales</td>\n",
       "      <td>Pyrodictiaceae</td>\n",
       "      <td>Pyrolobus</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Sulfolobales</td>\n",
       "      <td>Sulfolobaceae</td>\n",
       "      <td>Stygiolobus</td>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Thermoproteales</td>\n",
       "      <td>Thermofilaceae</td>\n",
       "      <td>Thermofilum</td>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomicrobia</td>\n",
       "      <td>Methanocellales</td>\n",
       "      <td>Methanocellaceae</td>\n",
       "      <td>Methanocella</td>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomicrobia</td>\n",
       "      <td>Methanosarcinales</td>\n",
       "      <td>Methanosarcinaceae</td>\n",
       "      <td>Methanimicrococcus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomicrobia</td>\n",
       "      <td>Methanosarcinales</td>\n",
       "      <td>Methermicoccaceae</td>\n",
       "      <td>Methermicoccus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Archaeoglobales</td>\n",
       "      <td>Archaeoglobaceae</td>\n",
       "      <td>Ferroglobus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Archaeoglobales</td>\n",
       "      <td>Archaeoglobaceae</td>\n",
       "      <td>Geoglobus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Haloplanus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Haloquadratum</td>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Halosimplex</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Natronobacterium</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Natronomonas</td>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanococci</td>\n",
       "      <td>Methanococcales</td>\n",
       "      <td>Methanocaldococcaceae</td>\n",
       "      <td>Methanotorris</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanopyri</td>\n",
       "      <td>Methanopyrales</td>\n",
       "      <td>Methanopyraceae</td>\n",
       "      <td>Methanopyrus</td>\n",
       "      <td>12</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Thermoplasmatales</td>\n",
       "      <td>Thermoplasmatales_incertae_sedis</td>\n",
       "      <td>Thermogymnomonas</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Acidothermaceae</td>\n",
       "      <td>Acidothermus</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Beutenbergiaceae</td>\n",
       "      <td>Salana</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Bogoriellaceae</td>\n",
       "      <td>Bogoriella</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Cellulomonadaceae</td>\n",
       "      <td>Tropheryma</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Corynebacteriaceae</td>\n",
       "      <td>Turicella</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Cryptosporangiaceae</td>\n",
       "      <td>Cryptosporangium</td>\n",
       "      <td>53</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Dermabacteraceae</td>\n",
       "      <td>Dermabacter</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Dermabacteraceae</td>\n",
       "      <td>Devriesea</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Dermatophilaceae</td>\n",
       "      <td>Kineosphaera</td>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Frankineae_incertae_sedis</td>\n",
       "      <td>Fodinicola</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Geodermatophilaceae</td>\n",
       "      <td>Blastococcus</td>\n",
       "      <td>99</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Geodermatophilaceae</td>\n",
       "      <td>Geodermatophilus</td>\n",
       "      <td>17</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Intrasporangiaceae</td>\n",
       "      <td>Kribbia</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2366</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oceanospirillaceae</td>\n",
       "      <td>Neptuniibacter</td>\n",
       "      <td>22</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2367</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oceanospirillales_incertae_sedis</td>\n",
       "      <td>Spongiispira</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2368</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oleiphilaceae</td>\n",
       "      <td>Oleiphilus</td>\n",
       "      <td>4</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2369</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pasteurellales</td>\n",
       "      <td>Pasteurellaceae</td>\n",
       "      <td>Haemophilus</td>\n",
       "      <td>5</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2370</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Alkanindiges</td>\n",
       "      <td>6</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2371</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Perlucidibaca</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2372</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Azomonas</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2373</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Azorhizophilus</td>\n",
       "      <td>2</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2374</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Azotobacter</td>\n",
       "      <td>2</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2376</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Piscirickettsiaceae</td>\n",
       "      <td>Cycloclasticus</td>\n",
       "      <td>2</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Piscirickettsiaceae</td>\n",
       "      <td>Sulfurivirga</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Thiotrichaceae</td>\n",
       "      <td>Beggiatoa</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Thiotrichales_incertae_sedis</td>\n",
       "      <td>Fangia</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Sinobacteraceae</td>\n",
       "      <td>Alkanibacter</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2381</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Sinobacteraceae</td>\n",
       "      <td>Sinobacter</td>\n",
       "      <td>7</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2382</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Aquimonas</td>\n",
       "      <td>9</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2383</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Frateuria</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2384</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Ignatzschineria</td>\n",
       "      <td>2</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2385</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Luteibacter</td>\n",
       "      <td>85</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2386</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Synergistetes</td>\n",
       "      <td>Synergistia</td>\n",
       "      <td>Synergistales</td>\n",
       "      <td>Synergistaceae</td>\n",
       "      <td>Aminomonas</td>\n",
       "      <td>1388</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2387</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>Mollicutes</td>\n",
       "      <td>Anaeroplasmatales</td>\n",
       "      <td>Anaeroplasmataceae</td>\n",
       "      <td>Asteroleplasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>Mollicutes</td>\n",
       "      <td>Haloplasmatales</td>\n",
       "      <td>Haloplasmataceae</td>\n",
       "      <td>Haloplasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteriales</td>\n",
       "      <td>Thermodesulfobacteriaceae</td>\n",
       "      <td>Caldimicrobium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2390</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteriales</td>\n",
       "      <td>Thermodesulfobacteriaceae</td>\n",
       "      <td>Thermodesulfatator</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2391</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogales</td>\n",
       "      <td>Thermotogaceae</td>\n",
       "      <td>Geotoga</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2392</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogales</td>\n",
       "      <td>Thermotogaceae</td>\n",
       "      <td>Kosmotoga</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Verrucomicrobia</td>\n",
       "      <td>Opitutae</td>\n",
       "      <td>Opitutales</td>\n",
       "      <td>Opitutaceae</td>\n",
       "      <td>Opitutus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Chlorarachniophyceae</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2396 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        domain                 phylum                  class  \\\n",
       "0      Archaea          Crenarchaeota           Thermoprotei   \n",
       "1      Archaea          Crenarchaeota           Thermoprotei   \n",
       "2      Archaea          Crenarchaeota           Thermoprotei   \n",
       "3      Archaea          Crenarchaeota           Thermoprotei   \n",
       "4      Archaea          Euryarchaeota        Methanomicrobia   \n",
       "5      Archaea          Euryarchaeota        Methanomicrobia   \n",
       "6      Archaea          Euryarchaeota        Methanomicrobia   \n",
       "7      Archaea          Euryarchaeota           Archaeoglobi   \n",
       "8      Archaea          Euryarchaeota           Archaeoglobi   \n",
       "9      Archaea          Euryarchaeota           Halobacteria   \n",
       "10     Archaea          Euryarchaeota           Halobacteria   \n",
       "11     Archaea          Euryarchaeota           Halobacteria   \n",
       "12     Archaea          Euryarchaeota           Halobacteria   \n",
       "13     Archaea          Euryarchaeota           Halobacteria   \n",
       "14     Archaea          Euryarchaeota           Methanococci   \n",
       "15     Archaea          Euryarchaeota            Methanopyri   \n",
       "16     Archaea          Euryarchaeota         Thermoplasmata   \n",
       "17    Bacteria         Actinobacteria         Actinobacteria   \n",
       "18    Bacteria         Actinobacteria         Actinobacteria   \n",
       "19    Bacteria         Actinobacteria         Actinobacteria   \n",
       "20    Bacteria         Actinobacteria         Actinobacteria   \n",
       "21    Bacteria         Actinobacteria         Actinobacteria   \n",
       "22    Bacteria         Actinobacteria         Actinobacteria   \n",
       "23    Bacteria         Actinobacteria         Actinobacteria   \n",
       "24    Bacteria         Actinobacteria         Actinobacteria   \n",
       "25    Bacteria         Actinobacteria         Actinobacteria   \n",
       "26    Bacteria         Actinobacteria         Actinobacteria   \n",
       "27    Bacteria         Actinobacteria         Actinobacteria   \n",
       "28    Bacteria         Actinobacteria         Actinobacteria   \n",
       "29    Bacteria         Actinobacteria         Actinobacteria   \n",
       "...        ...                    ...                    ...   \n",
       "2366  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2367  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2368  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2369  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2370  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2371  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2372  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2373  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2374  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2375  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2376  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2377  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2378  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2379  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2380  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2381  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2382  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2383  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2384  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2385  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2386  Bacteria          Synergistetes            Synergistia   \n",
       "2387  Bacteria            Tenericutes             Mollicutes   \n",
       "2388  Bacteria            Tenericutes             Mollicutes   \n",
       "2389  Bacteria  Thermodesulfobacteria  Thermodesulfobacteria   \n",
       "2390  Bacteria  Thermodesulfobacteria  Thermodesulfobacteria   \n",
       "2391  Bacteria            Thermotogae            Thermotogae   \n",
       "2392  Bacteria            Thermotogae            Thermotogae   \n",
       "2393  Bacteria        Verrucomicrobia               Opitutae   \n",
       "2394  Bacteria          Cyanobacteria          Cyanobacteria   \n",
       "2395  Bacteria          Cyanobacteria          Cyanobacteria   \n",
       "\n",
       "                         order                            family  \\\n",
       "0            Desulfurococcales                Desulfurococcaceae   \n",
       "1            Desulfurococcales                    Pyrodictiaceae   \n",
       "2                 Sulfolobales                     Sulfolobaceae   \n",
       "3              Thermoproteales                    Thermofilaceae   \n",
       "4              Methanocellales                  Methanocellaceae   \n",
       "5            Methanosarcinales                Methanosarcinaceae   \n",
       "6            Methanosarcinales                 Methermicoccaceae   \n",
       "7              Archaeoglobales                  Archaeoglobaceae   \n",
       "8              Archaeoglobales                  Archaeoglobaceae   \n",
       "9              Halobacteriales                  Halobacteriaceae   \n",
       "10             Halobacteriales                  Halobacteriaceae   \n",
       "11             Halobacteriales                  Halobacteriaceae   \n",
       "12             Halobacteriales                  Halobacteriaceae   \n",
       "13             Halobacteriales                  Halobacteriaceae   \n",
       "14             Methanococcales             Methanocaldococcaceae   \n",
       "15              Methanopyrales                   Methanopyraceae   \n",
       "16           Thermoplasmatales  Thermoplasmatales_incertae_sedis   \n",
       "17             Actinomycetales                   Acidothermaceae   \n",
       "18             Actinomycetales                  Beutenbergiaceae   \n",
       "19             Actinomycetales                    Bogoriellaceae   \n",
       "20             Actinomycetales                 Cellulomonadaceae   \n",
       "21             Actinomycetales                Corynebacteriaceae   \n",
       "22             Actinomycetales               Cryptosporangiaceae   \n",
       "23             Actinomycetales                  Dermabacteraceae   \n",
       "24             Actinomycetales                  Dermabacteraceae   \n",
       "25             Actinomycetales                  Dermatophilaceae   \n",
       "26             Actinomycetales         Frankineae_incertae_sedis   \n",
       "27             Actinomycetales               Geodermatophilaceae   \n",
       "28             Actinomycetales               Geodermatophilaceae   \n",
       "29             Actinomycetales                Intrasporangiaceae   \n",
       "...                        ...                               ...   \n",
       "2366         Oceanospirillales                Oceanospirillaceae   \n",
       "2367         Oceanospirillales  Oceanospirillales_incertae_sedis   \n",
       "2368         Oceanospirillales                     Oleiphilaceae   \n",
       "2369            Pasteurellales                   Pasteurellaceae   \n",
       "2370           Pseudomonadales                     Moraxellaceae   \n",
       "2371           Pseudomonadales                     Moraxellaceae   \n",
       "2372           Pseudomonadales                  Pseudomonadaceae   \n",
       "2373           Pseudomonadales                  Pseudomonadaceae   \n",
       "2374           Pseudomonadales                  Pseudomonadaceae   \n",
       "2375           Pseudomonadales                  Pseudomonadaceae   \n",
       "2376             Thiotrichales               Piscirickettsiaceae   \n",
       "2377             Thiotrichales               Piscirickettsiaceae   \n",
       "2378             Thiotrichales                    Thiotrichaceae   \n",
       "2379             Thiotrichales      Thiotrichales_incertae_sedis   \n",
       "2380           Xanthomonadales                   Sinobacteraceae   \n",
       "2381           Xanthomonadales                   Sinobacteraceae   \n",
       "2382           Xanthomonadales                  Xanthomonadaceae   \n",
       "2383           Xanthomonadales                  Xanthomonadaceae   \n",
       "2384           Xanthomonadales                  Xanthomonadaceae   \n",
       "2385           Xanthomonadales                  Xanthomonadaceae   \n",
       "2386             Synergistales                    Synergistaceae   \n",
       "2387         Anaeroplasmatales                Anaeroplasmataceae   \n",
       "2388           Haloplasmatales                  Haloplasmataceae   \n",
       "2389  Thermodesulfobacteriales         Thermodesulfobacteriaceae   \n",
       "2390  Thermodesulfobacteriales         Thermodesulfobacteriaceae   \n",
       "2391             Thermotogales                    Thermotogaceae   \n",
       "2392             Thermotogales                    Thermotogaceae   \n",
       "2393                Opitutales                       Opitutaceae   \n",
       "2394               Chloroplast              Chlorarachniophyceae   \n",
       "2395               Chloroplast                      Streptophyta   \n",
       "\n",
       "                   genus value          group_type group_phase  sample  \n",
       "0            Ignisphaera     7  EXTRACTION CONTROL      unknow  unknow  \n",
       "1              Pyrolobus     2  EXTRACTION CONTROL      unknow  unknow  \n",
       "2            Stygiolobus     3  EXTRACTION CONTROL      unknow  unknow  \n",
       "3            Thermofilum     3  EXTRACTION CONTROL      unknow  unknow  \n",
       "4           Methanocella     7  EXTRACTION CONTROL      unknow  unknow  \n",
       "5     Methanimicrococcus     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "6         Methermicoccus     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "7            Ferroglobus     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "8              Geoglobus     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "9             Haloplanus     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "10         Haloquadratum     4  EXTRACTION CONTROL      unknow  unknow  \n",
       "11           Halosimplex     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "12      Natronobacterium     2  EXTRACTION CONTROL      unknow  unknow  \n",
       "13          Natronomonas     4  EXTRACTION CONTROL      unknow  unknow  \n",
       "14         Methanotorris     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "15          Methanopyrus    12  EXTRACTION CONTROL      unknow  unknow  \n",
       "16      Thermogymnomonas     2  EXTRACTION CONTROL      unknow  unknow  \n",
       "17          Acidothermus     2  EXTRACTION CONTROL      unknow  unknow  \n",
       "18                Salana     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "19            Bogoriella     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "20            Tropheryma     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "21             Turicella     2  EXTRACTION CONTROL      unknow  unknow  \n",
       "22      Cryptosporangium    53  EXTRACTION CONTROL      unknow  unknow  \n",
       "23           Dermabacter     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "24             Devriesea     2  EXTRACTION CONTROL      unknow  unknow  \n",
       "25          Kineosphaera     3  EXTRACTION CONTROL      unknow  unknow  \n",
       "26            Fodinicola     1  EXTRACTION CONTROL      unknow  unknow  \n",
       "27          Blastococcus    99  EXTRACTION CONTROL      unknow  unknow  \n",
       "28      Geodermatophilus    17  EXTRACTION CONTROL      unknow  unknow  \n",
       "29               Kribbia     2  EXTRACTION CONTROL      unknow  unknow  \n",
       "...                  ...   ...                 ...         ...     ...  \n",
       "2366      Neptuniibacter    22             Control           2   stool  \n",
       "2367        Spongiispira     1             Control           2   stool  \n",
       "2368          Oleiphilus     4             Control           2   stool  \n",
       "2369         Haemophilus     5             Control           2   stool  \n",
       "2370        Alkanindiges     6             Control           2   stool  \n",
       "2371       Perlucidibaca     1             Control           2   stool  \n",
       "2372            Azomonas     1             Control           2   stool  \n",
       "2373      Azorhizophilus     2             Control           2   stool  \n",
       "2374         Azotobacter     2             Control           2   stool  \n",
       "2375         Pseudomonas     1             Control           2   stool  \n",
       "2376      Cycloclasticus     2             Control           2   stool  \n",
       "2377        Sulfurivirga     1             Control           2   stool  \n",
       "2378           Beggiatoa     1             Control           2   stool  \n",
       "2379              Fangia     1             Control           2   stool  \n",
       "2380        Alkanibacter     1             Control           2   stool  \n",
       "2381          Sinobacter     7             Control           2   stool  \n",
       "2382           Aquimonas     9             Control           2   stool  \n",
       "2383           Frateuria     1             Control           2   stool  \n",
       "2384     Ignatzschineria     2             Control           2   stool  \n",
       "2385         Luteibacter    85             Control           2   stool  \n",
       "2386          Aminomonas  1388             Control           2   stool  \n",
       "2387      Asteroleplasma   NaN             Control           2   stool  \n",
       "2388          Haloplasma   NaN             Control           2   stool  \n",
       "2389      Caldimicrobium   NaN             Control           2   stool  \n",
       "2390  Thermodesulfatator   NaN             Control           2   stool  \n",
       "2391             Geotoga   NaN             Control           2   stool  \n",
       "2392           Kosmotoga   NaN             Control           2   stool  \n",
       "2393            Opitutus   NaN             Control           2   stool  \n",
       "2394                None   NaN             Control           2   stool  \n",
       "2395                None   NaN             Control           2   stool  \n",
       "\n",
       "[2396 rows x 10 columns]"
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loop over all the datasheets\n",
    "clean_data = pd.DataFrame(columns=df_col_list)\n",
    "for metadata_row in metadata.itertuples():\n",
    "    raw_data = pd.read_excel(MICROBIOME_FOLDER+\"/\"+metadata_row.barcode+\".xls\", sheetname='Sheet 1', header=None)\n",
    "    # Change column names to something clearer\n",
    "    raw_data.columns = [\"raw_classification\", \"value\"]\n",
    "    \n",
    "    # For each datasheet create a local classified set of data\n",
    "    classifier_array = []\n",
    "    for row_data in raw_data.itertuples():\n",
    "        classifier_array.append(get_classifiers(row_data.raw_classification))\n",
    "    local_classified = pd.DataFrame(classifier_array, columns=scientific_classification)\n",
    "    # To the local set of data add the columns that are not the classification\n",
    "    local_classified[\"value\"]       = test_data[\"value\"]\n",
    "    local_classified[\"group_type\"]  = metadata_row.group_type\n",
    "    local_classified[\"group_phase\"] = metadata_row.group_phase\n",
    "    local_classified[\"sample\"]      = metadata_row.sample\n",
    "    # Append this local set of data to \"clean_data\", which is the merge of the data from each measurements\n",
    "    clean_data = clean_data.append(local_classified, ignore_index=True)\n",
    "\n",
    "clean_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Now lets replace all the None values by unknow **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>value</th>\n",
       "      <th>group_type</th>\n",
       "      <th>group_phase</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Desulfurococcales</td>\n",
       "      <td>Desulfurococcaceae</td>\n",
       "      <td>Ignisphaera</td>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Desulfurococcales</td>\n",
       "      <td>Pyrodictiaceae</td>\n",
       "      <td>Pyrolobus</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Sulfolobales</td>\n",
       "      <td>Sulfolobaceae</td>\n",
       "      <td>Stygiolobus</td>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Thermoprotei</td>\n",
       "      <td>Thermoproteales</td>\n",
       "      <td>Thermofilaceae</td>\n",
       "      <td>Thermofilum</td>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomicrobia</td>\n",
       "      <td>Methanocellales</td>\n",
       "      <td>Methanocellaceae</td>\n",
       "      <td>Methanocella</td>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomicrobia</td>\n",
       "      <td>Methanosarcinales</td>\n",
       "      <td>Methanosarcinaceae</td>\n",
       "      <td>Methanimicrococcus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomicrobia</td>\n",
       "      <td>Methanosarcinales</td>\n",
       "      <td>Methermicoccaceae</td>\n",
       "      <td>Methermicoccus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Archaeoglobales</td>\n",
       "      <td>Archaeoglobaceae</td>\n",
       "      <td>Ferroglobus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Archaeoglobi</td>\n",
       "      <td>Archaeoglobales</td>\n",
       "      <td>Archaeoglobaceae</td>\n",
       "      <td>Geoglobus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Haloplanus</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Haloquadratum</td>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Halosimplex</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Natronobacterium</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Halobacteriaceae</td>\n",
       "      <td>Natronomonas</td>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanococci</td>\n",
       "      <td>Methanococcales</td>\n",
       "      <td>Methanocaldococcaceae</td>\n",
       "      <td>Methanotorris</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanopyri</td>\n",
       "      <td>Methanopyrales</td>\n",
       "      <td>Methanopyraceae</td>\n",
       "      <td>Methanopyrus</td>\n",
       "      <td>12</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Thermoplasmatales</td>\n",
       "      <td>Thermoplasmatales_incertae_sedis</td>\n",
       "      <td>Thermogymnomonas</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Acidothermaceae</td>\n",
       "      <td>Acidothermus</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Beutenbergiaceae</td>\n",
       "      <td>Salana</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Bogoriellaceae</td>\n",
       "      <td>Bogoriella</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Cellulomonadaceae</td>\n",
       "      <td>Tropheryma</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Corynebacteriaceae</td>\n",
       "      <td>Turicella</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Cryptosporangiaceae</td>\n",
       "      <td>Cryptosporangium</td>\n",
       "      <td>53</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Dermabacteraceae</td>\n",
       "      <td>Dermabacter</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Dermabacteraceae</td>\n",
       "      <td>Devriesea</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Dermatophilaceae</td>\n",
       "      <td>Kineosphaera</td>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Frankineae_incertae_sedis</td>\n",
       "      <td>Fodinicola</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Geodermatophilaceae</td>\n",
       "      <td>Blastococcus</td>\n",
       "      <td>99</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Geodermatophilaceae</td>\n",
       "      <td>Geodermatophilus</td>\n",
       "      <td>17</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinomycetales</td>\n",
       "      <td>Intrasporangiaceae</td>\n",
       "      <td>Kribbia</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2366</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oceanospirillaceae</td>\n",
       "      <td>Neptuniibacter</td>\n",
       "      <td>22</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2367</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oceanospirillales_incertae_sedis</td>\n",
       "      <td>Spongiispira</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2368</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>Oleiphilaceae</td>\n",
       "      <td>Oleiphilus</td>\n",
       "      <td>4</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2369</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pasteurellales</td>\n",
       "      <td>Pasteurellaceae</td>\n",
       "      <td>Haemophilus</td>\n",
       "      <td>5</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2370</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Alkanindiges</td>\n",
       "      <td>6</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2371</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Perlucidibaca</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2372</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Azomonas</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2373</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Azorhizophilus</td>\n",
       "      <td>2</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2374</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Azotobacter</td>\n",
       "      <td>2</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2376</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Piscirickettsiaceae</td>\n",
       "      <td>Cycloclasticus</td>\n",
       "      <td>2</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Piscirickettsiaceae</td>\n",
       "      <td>Sulfurivirga</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Thiotrichaceae</td>\n",
       "      <td>Beggiatoa</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Thiotrichales_incertae_sedis</td>\n",
       "      <td>Fangia</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Sinobacteraceae</td>\n",
       "      <td>Alkanibacter</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2381</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Sinobacteraceae</td>\n",
       "      <td>Sinobacter</td>\n",
       "      <td>7</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2382</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Aquimonas</td>\n",
       "      <td>9</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2383</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Frateuria</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2384</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Ignatzschineria</td>\n",
       "      <td>2</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2385</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Luteibacter</td>\n",
       "      <td>85</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2386</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Synergistetes</td>\n",
       "      <td>Synergistia</td>\n",
       "      <td>Synergistales</td>\n",
       "      <td>Synergistaceae</td>\n",
       "      <td>Aminomonas</td>\n",
       "      <td>1388</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2387</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>Mollicutes</td>\n",
       "      <td>Anaeroplasmatales</td>\n",
       "      <td>Anaeroplasmataceae</td>\n",
       "      <td>Asteroleplasma</td>\n",
       "      <td>unknow</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Tenericutes</td>\n",
       "      <td>Mollicutes</td>\n",
       "      <td>Haloplasmatales</td>\n",
       "      <td>Haloplasmataceae</td>\n",
       "      <td>Haloplasma</td>\n",
       "      <td>unknow</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteriales</td>\n",
       "      <td>Thermodesulfobacteriaceae</td>\n",
       "      <td>Caldimicrobium</td>\n",
       "      <td>unknow</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2390</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteria</td>\n",
       "      <td>Thermodesulfobacteriales</td>\n",
       "      <td>Thermodesulfobacteriaceae</td>\n",
       "      <td>Thermodesulfatator</td>\n",
       "      <td>unknow</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2391</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogales</td>\n",
       "      <td>Thermotogaceae</td>\n",
       "      <td>Geotoga</td>\n",
       "      <td>unknow</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2392</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogales</td>\n",
       "      <td>Thermotogaceae</td>\n",
       "      <td>Kosmotoga</td>\n",
       "      <td>unknow</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Verrucomicrobia</td>\n",
       "      <td>Opitutae</td>\n",
       "      <td>Opitutales</td>\n",
       "      <td>Opitutaceae</td>\n",
       "      <td>Opitutus</td>\n",
       "      <td>unknow</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Chlorarachniophyceae</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>unknow</td>\n",
       "      <td>unknow</td>\n",
       "      <td>Control</td>\n",
       "      <td>2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2396 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        domain                 phylum                  class  \\\n",
       "0      Archaea          Crenarchaeota           Thermoprotei   \n",
       "1      Archaea          Crenarchaeota           Thermoprotei   \n",
       "2      Archaea          Crenarchaeota           Thermoprotei   \n",
       "3      Archaea          Crenarchaeota           Thermoprotei   \n",
       "4      Archaea          Euryarchaeota        Methanomicrobia   \n",
       "5      Archaea          Euryarchaeota        Methanomicrobia   \n",
       "6      Archaea          Euryarchaeota        Methanomicrobia   \n",
       "7      Archaea          Euryarchaeota           Archaeoglobi   \n",
       "8      Archaea          Euryarchaeota           Archaeoglobi   \n",
       "9      Archaea          Euryarchaeota           Halobacteria   \n",
       "10     Archaea          Euryarchaeota           Halobacteria   \n",
       "11     Archaea          Euryarchaeota           Halobacteria   \n",
       "12     Archaea          Euryarchaeota           Halobacteria   \n",
       "13     Archaea          Euryarchaeota           Halobacteria   \n",
       "14     Archaea          Euryarchaeota           Methanococci   \n",
       "15     Archaea          Euryarchaeota            Methanopyri   \n",
       "16     Archaea          Euryarchaeota         Thermoplasmata   \n",
       "17    Bacteria         Actinobacteria         Actinobacteria   \n",
       "18    Bacteria         Actinobacteria         Actinobacteria   \n",
       "19    Bacteria         Actinobacteria         Actinobacteria   \n",
       "20    Bacteria         Actinobacteria         Actinobacteria   \n",
       "21    Bacteria         Actinobacteria         Actinobacteria   \n",
       "22    Bacteria         Actinobacteria         Actinobacteria   \n",
       "23    Bacteria         Actinobacteria         Actinobacteria   \n",
       "24    Bacteria         Actinobacteria         Actinobacteria   \n",
       "25    Bacteria         Actinobacteria         Actinobacteria   \n",
       "26    Bacteria         Actinobacteria         Actinobacteria   \n",
       "27    Bacteria         Actinobacteria         Actinobacteria   \n",
       "28    Bacteria         Actinobacteria         Actinobacteria   \n",
       "29    Bacteria         Actinobacteria         Actinobacteria   \n",
       "...        ...                    ...                    ...   \n",
       "2366  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2367  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2368  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2369  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2370  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2371  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2372  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2373  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2374  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2375  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2376  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2377  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2378  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2379  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2380  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2381  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2382  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2383  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2384  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2385  Bacteria         Proteobacteria    Gammaproteobacteria   \n",
       "2386  Bacteria          Synergistetes            Synergistia   \n",
       "2387  Bacteria            Tenericutes             Mollicutes   \n",
       "2388  Bacteria            Tenericutes             Mollicutes   \n",
       "2389  Bacteria  Thermodesulfobacteria  Thermodesulfobacteria   \n",
       "2390  Bacteria  Thermodesulfobacteria  Thermodesulfobacteria   \n",
       "2391  Bacteria            Thermotogae            Thermotogae   \n",
       "2392  Bacteria            Thermotogae            Thermotogae   \n",
       "2393  Bacteria        Verrucomicrobia               Opitutae   \n",
       "2394  Bacteria          Cyanobacteria          Cyanobacteria   \n",
       "2395  Bacteria          Cyanobacteria          Cyanobacteria   \n",
       "\n",
       "                         order                            family  \\\n",
       "0            Desulfurococcales                Desulfurococcaceae   \n",
       "1            Desulfurococcales                    Pyrodictiaceae   \n",
       "2                 Sulfolobales                     Sulfolobaceae   \n",
       "3              Thermoproteales                    Thermofilaceae   \n",
       "4              Methanocellales                  Methanocellaceae   \n",
       "5            Methanosarcinales                Methanosarcinaceae   \n",
       "6            Methanosarcinales                 Methermicoccaceae   \n",
       "7              Archaeoglobales                  Archaeoglobaceae   \n",
       "8              Archaeoglobales                  Archaeoglobaceae   \n",
       "9              Halobacteriales                  Halobacteriaceae   \n",
       "10             Halobacteriales                  Halobacteriaceae   \n",
       "11             Halobacteriales                  Halobacteriaceae   \n",
       "12             Halobacteriales                  Halobacteriaceae   \n",
       "13             Halobacteriales                  Halobacteriaceae   \n",
       "14             Methanococcales             Methanocaldococcaceae   \n",
       "15              Methanopyrales                   Methanopyraceae   \n",
       "16           Thermoplasmatales  Thermoplasmatales_incertae_sedis   \n",
       "17             Actinomycetales                   Acidothermaceae   \n",
       "18             Actinomycetales                  Beutenbergiaceae   \n",
       "19             Actinomycetales                    Bogoriellaceae   \n",
       "20             Actinomycetales                 Cellulomonadaceae   \n",
       "21             Actinomycetales                Corynebacteriaceae   \n",
       "22             Actinomycetales               Cryptosporangiaceae   \n",
       "23             Actinomycetales                  Dermabacteraceae   \n",
       "24             Actinomycetales                  Dermabacteraceae   \n",
       "25             Actinomycetales                  Dermatophilaceae   \n",
       "26             Actinomycetales         Frankineae_incertae_sedis   \n",
       "27             Actinomycetales               Geodermatophilaceae   \n",
       "28             Actinomycetales               Geodermatophilaceae   \n",
       "29             Actinomycetales                Intrasporangiaceae   \n",
       "...                        ...                               ...   \n",
       "2366         Oceanospirillales                Oceanospirillaceae   \n",
       "2367         Oceanospirillales  Oceanospirillales_incertae_sedis   \n",
       "2368         Oceanospirillales                     Oleiphilaceae   \n",
       "2369            Pasteurellales                   Pasteurellaceae   \n",
       "2370           Pseudomonadales                     Moraxellaceae   \n",
       "2371           Pseudomonadales                     Moraxellaceae   \n",
       "2372           Pseudomonadales                  Pseudomonadaceae   \n",
       "2373           Pseudomonadales                  Pseudomonadaceae   \n",
       "2374           Pseudomonadales                  Pseudomonadaceae   \n",
       "2375           Pseudomonadales                  Pseudomonadaceae   \n",
       "2376             Thiotrichales               Piscirickettsiaceae   \n",
       "2377             Thiotrichales               Piscirickettsiaceae   \n",
       "2378             Thiotrichales                    Thiotrichaceae   \n",
       "2379             Thiotrichales      Thiotrichales_incertae_sedis   \n",
       "2380           Xanthomonadales                   Sinobacteraceae   \n",
       "2381           Xanthomonadales                   Sinobacteraceae   \n",
       "2382           Xanthomonadales                  Xanthomonadaceae   \n",
       "2383           Xanthomonadales                  Xanthomonadaceae   \n",
       "2384           Xanthomonadales                  Xanthomonadaceae   \n",
       "2385           Xanthomonadales                  Xanthomonadaceae   \n",
       "2386             Synergistales                    Synergistaceae   \n",
       "2387         Anaeroplasmatales                Anaeroplasmataceae   \n",
       "2388           Haloplasmatales                  Haloplasmataceae   \n",
       "2389  Thermodesulfobacteriales         Thermodesulfobacteriaceae   \n",
       "2390  Thermodesulfobacteriales         Thermodesulfobacteriaceae   \n",
       "2391             Thermotogales                    Thermotogaceae   \n",
       "2392             Thermotogales                    Thermotogaceae   \n",
       "2393                Opitutales                       Opitutaceae   \n",
       "2394               Chloroplast              Chlorarachniophyceae   \n",
       "2395               Chloroplast                      Streptophyta   \n",
       "\n",
       "                   genus   value          group_type group_phase  sample  \n",
       "0            Ignisphaera       7  EXTRACTION CONTROL      unknow  unknow  \n",
       "1              Pyrolobus       2  EXTRACTION CONTROL      unknow  unknow  \n",
       "2            Stygiolobus       3  EXTRACTION CONTROL      unknow  unknow  \n",
       "3            Thermofilum       3  EXTRACTION CONTROL      unknow  unknow  \n",
       "4           Methanocella       7  EXTRACTION CONTROL      unknow  unknow  \n",
       "5     Methanimicrococcus       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "6         Methermicoccus       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "7            Ferroglobus       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "8              Geoglobus       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "9             Haloplanus       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "10         Haloquadratum       4  EXTRACTION CONTROL      unknow  unknow  \n",
       "11           Halosimplex       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "12      Natronobacterium       2  EXTRACTION CONTROL      unknow  unknow  \n",
       "13          Natronomonas       4  EXTRACTION CONTROL      unknow  unknow  \n",
       "14         Methanotorris       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "15          Methanopyrus      12  EXTRACTION CONTROL      unknow  unknow  \n",
       "16      Thermogymnomonas       2  EXTRACTION CONTROL      unknow  unknow  \n",
       "17          Acidothermus       2  EXTRACTION CONTROL      unknow  unknow  \n",
       "18                Salana       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "19            Bogoriella       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "20            Tropheryma       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "21             Turicella       2  EXTRACTION CONTROL      unknow  unknow  \n",
       "22      Cryptosporangium      53  EXTRACTION CONTROL      unknow  unknow  \n",
       "23           Dermabacter       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "24             Devriesea       2  EXTRACTION CONTROL      unknow  unknow  \n",
       "25          Kineosphaera       3  EXTRACTION CONTROL      unknow  unknow  \n",
       "26            Fodinicola       1  EXTRACTION CONTROL      unknow  unknow  \n",
       "27          Blastococcus      99  EXTRACTION CONTROL      unknow  unknow  \n",
       "28      Geodermatophilus      17  EXTRACTION CONTROL      unknow  unknow  \n",
       "29               Kribbia       2  EXTRACTION CONTROL      unknow  unknow  \n",
       "...                  ...     ...                 ...         ...     ...  \n",
       "2366      Neptuniibacter      22             Control           2   stool  \n",
       "2367        Spongiispira       1             Control           2   stool  \n",
       "2368          Oleiphilus       4             Control           2   stool  \n",
       "2369         Haemophilus       5             Control           2   stool  \n",
       "2370        Alkanindiges       6             Control           2   stool  \n",
       "2371       Perlucidibaca       1             Control           2   stool  \n",
       "2372            Azomonas       1             Control           2   stool  \n",
       "2373      Azorhizophilus       2             Control           2   stool  \n",
       "2374         Azotobacter       2             Control           2   stool  \n",
       "2375         Pseudomonas       1             Control           2   stool  \n",
       "2376      Cycloclasticus       2             Control           2   stool  \n",
       "2377        Sulfurivirga       1             Control           2   stool  \n",
       "2378           Beggiatoa       1             Control           2   stool  \n",
       "2379              Fangia       1             Control           2   stool  \n",
       "2380        Alkanibacter       1             Control           2   stool  \n",
       "2381          Sinobacter       7             Control           2   stool  \n",
       "2382           Aquimonas       9             Control           2   stool  \n",
       "2383           Frateuria       1             Control           2   stool  \n",
       "2384     Ignatzschineria       2             Control           2   stool  \n",
       "2385         Luteibacter      85             Control           2   stool  \n",
       "2386          Aminomonas    1388             Control           2   stool  \n",
       "2387      Asteroleplasma  unknow             Control           2   stool  \n",
       "2388          Haloplasma  unknow             Control           2   stool  \n",
       "2389      Caldimicrobium  unknow             Control           2   stool  \n",
       "2390  Thermodesulfatator  unknow             Control           2   stool  \n",
       "2391             Geotoga  unknow             Control           2   stool  \n",
       "2392           Kosmotoga  unknow             Control           2   stool  \n",
       "2393            Opitutus  unknow             Control           2   stool  \n",
       "2394              unknow  unknow             Control           2   stool  \n",
       "2395              unknow  unknow             Control           2   stool  \n",
       "\n",
       "[2396 rows x 10 columns]"
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for item in clean_data.iteritems():\n",
    "    item[1].fillna(value=\"unknow\", inplace=True)\n",
    "clean_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
